// @aries-framework/core/build/modules/credentials/CredentialsModule.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CredentialsModule = void 0;
const models_1 = require("../../agent/models");
const CredentialsApi_1 = require("./CredentialsApi");
const CredentialsModuleConfig_1 = require("./CredentialsModuleConfig");
const indy_1 = require("./formats/indy");
const services_1 = require("./protocol/revocation-notification/services");
const v1_1 = require("./protocol/v1");
const v2_1 = require("./protocol/v2");
const repository_1 = require("./repository");
class CredentialsModule {
    constructor(config) {
        var _a;
        // Infer Api type from the config
        this.api = CredentialsApi_1.CredentialsApi;
        this.config = new CredentialsModuleConfig_1.CredentialsModuleConfig(Object.assign(Object.assign({}, config), { 
            // NOTE: the credentialProtocols defaults are set in the CredentialsModule rather than the CredentialsModuleConfig to
            // void dependency cycles.
            credentialProtocols: (_a = config === null || config === void 0 ? void 0 : config.credentialProtocols) !== null && _a !== void 0 ? _a : this.getDefaultCredentialProtocols() }));
    }
    /**
     * Get the default credential protocols that will be registered if the `credentialProtocols` property is not configured.
     */
    getDefaultCredentialProtocols() {
        // Instantiate credential formats
        const indyCredentialFormat = new indy_1.IndyCredentialFormatService();
        // Instantiate credential protocols
        const v1CredentialProtocol = new v1_1.V1CredentialProtocol({ indyCredentialFormat });
        const v2CredentialProtocol = new v2_1.V2CredentialProtocol({
            credentialFormats: [indyCredentialFormat],
        });
        return [v1CredentialProtocol, v2CredentialProtocol];
    }
    /**
     * Registers the dependencies of the credentials module on the dependency manager.
     */
    register(dependencyManager, featureRegistry) {
        // Api
        dependencyManager.registerContextScoped(CredentialsApi_1.CredentialsApi);
        // Config
        dependencyManager.registerInstance(CredentialsModuleConfig_1.CredentialsModuleConfig, this.config);
        // Services
        dependencyManager.registerSingleton(services_1.RevocationNotificationService);
        // Repositories
        dependencyManager.registerSingleton(repository_1.CredentialRepository);
        // Features
        featureRegistry.register(new models_1.Protocol({
            id: 'https://didcomm.org/revocation_notification/1.0',
            roles: ['holder'],
        }), new models_1.Protocol({
            id: 'https://didcomm.org/revocation_notification/2.0',
            roles: ['holder'],
        }));
        // Protocol needs to register feature registry items and handlers
        for (const credentialProtocol of this.config.credentialProtocols) {
            credentialProtocol.register(dependencyManager, featureRegistry);
        }
    }
}
exports.CredentialsModule = CredentialsModule;
