// @react-native-windows/cli/lib-commonjs/healthChecks.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getHealthChecks = void 0;
const path_1 = __importDefault(require("path"));
const rnDir = path_1.default.dirname(require.resolve('react-native/package.json', { paths: [process.cwd()] }));
const cliDir = path_1.default.dirname(require.resolve('@react-native-community/cli/package.json', { paths: [rnDir] }));
const execaPath = require.resolve('execa', { paths: [cliDir] });
const execa = require(execaPath);
const commandWithProgress_1 = require("./runWindows/utils/commandWithProgress");
const healthCheckList_1 = require("./healthCheckList");
function getHealthChecks() {
    // #8471: There are known cases where the dependencies script will error out.
    // Fail gracefully if that happens in the meantime.
    try {
        return getHealthChecksUnsafe();
    }
    catch (_a) {
        return [{
                label: 'Windows',
                healthchecks: [{
                        label: 'Failed to enumerate health checks',
                        description: 'Failed to enumerate health checks',
                        getDiagnostics: async () => ({ needsToBeFixed: true }),
                        runAutomaticFix: async ({ loader }) => { loader.fail(); },
                    }]
            }];
    }
}
exports.getHealthChecks = getHealthChecks;
function getHealthChecksUnsafe() {
    // All our health checks are windows only...
    if (process.platform !== 'win32') {
        return undefined;
    }
    const rnwDepScriptPath = path_1.default.join(path_1.default.dirname(require.resolve('react-native-windows/package.json', { paths: [process.cwd()] })), 'Scripts/rnw-dependencies.ps1');
    return [
        {
            label: 'Windows',
            healthchecks: healthCheckList_1.HealthCheckList.map((dep) => {
                const [optional, id, name] = dep;
                return {
                    label: name,
                    description: name,
                    // The schema check of react-native doctor requires this to be a string, although it should be a boolean
                    // Enable this once we pick up a new version of the CLI that includes https://github.com/react-native-community/cli/pull/1367
                    isRequired: optional,
                    getDiagnostics: async () => {
                        let needsToBeFixed = true;
                        try {
                            await execa(`"${commandWithProgress_1.powershell}" -ExecutionPolicy Unrestricted -NoProfile "${rnwDepScriptPath}" -NoPrompt -Check ${id}`);
                            needsToBeFixed = false;
                        }
                        catch (_a) {
                        }
                        return {
                            needsToBeFixed,
                        };
                    },
                    runAutomaticFix: async ({ loader, logManualInstallation }) => {
                        const command = `${commandWithProgress_1.powershell} -ExecutionPolicy Unrestricted -NoProfile "${rnwDepScriptPath}" -Check ${id}`;
                        try {
                            const { exitCode } = await execa(command, { stdio: 'inherit' });
                            if (exitCode) {
                                logManualInstallation({ command, healthcheck: `react-native-windows dependency "${id}"` });
                                loader.fail();
                            }
                            else {
                                loader.succeed();
                            }
                        }
                        catch (_a) {
                            logManualInstallation({ command, healthcheck: `react-native-windows dependency "${id}"` });
                            loader.fail();
                        }
                    }
                };
            })
        }
    ];
}
