// @aries-framework/core/build/modules/proofs/protocol/v1/messages/V1RequestPresentationMessage.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.V1RequestPresentationMessage = exports.INDY_PROOF_REQUEST_ATTACHMENT_ID = void 0;
const class_transformer_1 = require("class-transformer");
const class_validator_1 = require("class-validator");
const AgentMessage_1 = require("../../../../../agent/AgentMessage");
const Attachment_1 = require("../../../../../decorators/attachment/Attachment");
const AriesFrameworkError_1 = require("../../../../../error/AriesFrameworkError");
const JsonTransformer_1 = require("../../../../../utils/JsonTransformer");
const messageType_1 = require("../../../../../utils/messageType");
const ProofFormatConstants_1 = require("../../../formats/ProofFormatConstants");
const ProofRequest_1 = require("../../../formats/indy/models/ProofRequest");
const ProofFormatSpec_1 = require("../../../models/ProofFormatSpec");
exports.INDY_PROOF_REQUEST_ATTACHMENT_ID = 'libindy-request-presentation-0';
/**
 * Request Presentation Message part of Present Proof Protocol used to initiate request from verifier to prover.
 *
 * @see https://github.com/hyperledger/aries-rfcs/blob/master/features/0037-present-proof/README.md#request-presentation
 */
class V1RequestPresentationMessage extends AgentMessage_1.AgentMessage {
    constructor(options) {
        var _a;
        super();
        this.type = V1RequestPresentationMessage.type.messageTypeUri;
        if (options) {
            this.id = (_a = options.id) !== null && _a !== void 0 ? _a : this.generateId();
            this.comment = options.comment;
            this.requestPresentationAttachments = options.requestPresentationAttachments;
            if (options.parentThreadId) {
                this.setThread({
                    parentThreadId: options.parentThreadId,
                });
            }
        }
    }
    get indyProofRequest() {
        // Extract proof request from attachment
        const proofRequestJson = this.indyProofRequestJson;
        const proofRequest = JsonTransformer_1.JsonTransformer.fromJSON(proofRequestJson, ProofRequest_1.ProofRequest);
        return proofRequest;
    }
    get indyProofRequestJson() {
        var _a;
        const attachment = this.requestPresentationAttachments.find((attachment) => attachment.id === exports.INDY_PROOF_REQUEST_ATTACHMENT_ID);
        // Extract proof request from attachment
        return (_a = attachment === null || attachment === void 0 ? void 0 : attachment.getDataAsJson()) !== null && _a !== void 0 ? _a : null;
    }
    getAttachmentFormats() {
        const attachment = this.indyAttachment;
        if (!attachment) {
            throw new AriesFrameworkError_1.AriesFrameworkError(`Could not find a request presentation attachment`);
        }
        return [
            {
                format: new ProofFormatSpec_1.ProofFormatSpec({ format: ProofFormatConstants_1.V2_INDY_PRESENTATION_REQUEST }),
                attachment: attachment,
            },
        ];
    }
    get indyAttachment() {
        var _a;
        return ((_a = this.requestPresentationAttachments.find((attachment) => attachment.id === exports.INDY_PROOF_REQUEST_ATTACHMENT_ID)) !== null && _a !== void 0 ? _a : null);
    }
}
V1RequestPresentationMessage.type = (0, messageType_1.parseMessageType)('https://didcomm.org/present-proof/1.0/request-presentation');
__decorate([
    (0, messageType_1.IsValidMessageType)(V1RequestPresentationMessage.type),
    __metadata("design:type", Object)
], V1RequestPresentationMessage.prototype, "type", void 0);
__decorate([
    (0, class_validator_1.IsOptional)(),
    (0, class_validator_1.IsString)(),
    __metadata("design:type", String)
], V1RequestPresentationMessage.prototype, "comment", void 0);
__decorate([
    (0, class_transformer_1.Expose)({ name: 'request_presentations~attach' }),
    (0, class_transformer_1.Type)(() => Attachment_1.Attachment),
    (0, class_validator_1.IsArray)(),
    (0, class_validator_1.ValidateNested)({
        each: true,
    }),
    (0, class_validator_1.IsInstance)(Attachment_1.Attachment, { each: true }),
    __metadata("design:type", Array)
], V1RequestPresentationMessage.prototype, "requestPresentationAttachments", void 0);
exports.V1RequestPresentationMessage = V1RequestPresentationMessage;
