// @aries-framework/core/build/modules/proofs/formats/indy/models/ProofPredicateInfo.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ProofPredicateInfo = void 0;
const class_transformer_1 = require("class-transformer");
const class_validator_1 = require("class-validator");
const credentials_1 = require("../../../../credentials");
const AttributeFilter_1 = require("./AttributeFilter");
const PredicateType_1 = require("./PredicateType");
class ProofPredicateInfo {
    constructor(options) {
        if (options) {
            this.name = options.name;
            this.nonRevoked = options.nonRevoked;
            this.restrictions = options.restrictions;
            this.predicateType = options.predicateType;
            this.predicateValue = options.predicateValue;
        }
    }
}
__decorate([
    (0, class_validator_1.IsString)(),
    __metadata("design:type", String)
], ProofPredicateInfo.prototype, "name", void 0);
__decorate([
    (0, class_transformer_1.Expose)({ name: 'p_type' }),
    (0, class_validator_1.IsEnum)(PredicateType_1.PredicateType),
    __metadata("design:type", String)
], ProofPredicateInfo.prototype, "predicateType", void 0);
__decorate([
    (0, class_transformer_1.Expose)({ name: 'p_value' }),
    (0, class_validator_1.IsInt)(),
    __metadata("design:type", Number)
], ProofPredicateInfo.prototype, "predicateValue", void 0);
__decorate([
    (0, class_transformer_1.Expose)({ name: 'non_revoked' }),
    (0, class_validator_1.ValidateNested)(),
    (0, class_transformer_1.Type)(() => credentials_1.IndyRevocationInterval),
    (0, class_validator_1.IsOptional)(),
    (0, class_validator_1.IsInstance)(credentials_1.IndyRevocationInterval),
    __metadata("design:type", credentials_1.IndyRevocationInterval)
], ProofPredicateInfo.prototype, "nonRevoked", void 0);
__decorate([
    (0, class_validator_1.ValidateNested)({ each: true }),
    (0, class_transformer_1.Type)(() => AttributeFilter_1.AttributeFilter),
    (0, class_validator_1.IsOptional)(),
    (0, class_validator_1.IsInstance)(AttributeFilter_1.AttributeFilter, { each: true }),
    (0, class_validator_1.IsArray)(),
    __metadata("design:type", Array)
], ProofPredicateInfo.prototype, "restrictions", void 0);
exports.ProofPredicateInfo = ProofPredicateInfo;
