// @aries-framework/core/build/modules/proofs/formats/indy/models/RequestedAttribute.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.RequestedAttribute = void 0;
const class_transformer_1 = require("class-transformer");
const class_validator_1 = require("class-validator");
const IndyCredentialInfo_1 = require("../../../../credentials/formats/indy/models/IndyCredentialInfo");
/**
 * Requested Attribute for Indy proof creation
 */
class RequestedAttribute {
    constructor(options) {
        if (options) {
            this.credentialId = options.credentialId;
            this.timestamp = options.timestamp;
            this.revealed = options.revealed;
            this.credentialInfo = options.credentialInfo;
            this.revoked = options.revoked;
        }
    }
}
__decorate([
    (0, class_transformer_1.Expose)({ name: 'cred_id' }),
    (0, class_validator_1.IsString)(),
    __metadata("design:type", String)
], RequestedAttribute.prototype, "credentialId", void 0);
__decorate([
    (0, class_transformer_1.Expose)({ name: 'timestamp' }),
    (0, class_validator_1.IsInt)(),
    (0, class_validator_1.IsOptional)(),
    __metadata("design:type", Number)
], RequestedAttribute.prototype, "timestamp", void 0);
__decorate([
    (0, class_validator_1.IsBoolean)(),
    __metadata("design:type", Boolean)
], RequestedAttribute.prototype, "revealed", void 0);
__decorate([
    (0, class_transformer_1.Exclude)({ toPlainOnly: true }),
    __metadata("design:type", IndyCredentialInfo_1.IndyCredentialInfo)
], RequestedAttribute.prototype, "credentialInfo", void 0);
__decorate([
    (0, class_transformer_1.Exclude)({ toPlainOnly: true }),
    __metadata("design:type", Boolean)
], RequestedAttribute.prototype, "revoked", void 0);
exports.RequestedAttribute = RequestedAttribute;
