// @aries-framework/core/build/modules/ledger/services/IndyPoolService.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.IndyPoolService = exports.DID_POOL_CACHE_LIMIT = exports.DID_POOL_CACHE_ID = void 0;
const rxjs_1 = require("rxjs");
const cache_1 = require("../../../cache");
const constants_1 = require("../../../constants");
const IndySdkError_1 = require("../../../error/IndySdkError");
const plugins_1 = require("../../../plugins");
const did_1 = require("../../../utils/did");
const indyError_1 = require("../../../utils/indyError");
const promises_1 = require("../../../utils/promises");
const assertIndyWallet_1 = require("../../../wallet/util/assertIndyWallet");
const IndyPool_1 = require("../IndyPool");
const LedgerError_1 = require("../error/LedgerError");
const LedgerNotConfiguredError_1 = require("../error/LedgerNotConfiguredError");
const LedgerNotFoundError_1 = require("../error/LedgerNotFoundError");
exports.DID_POOL_CACHE_ID = 'DID_POOL_CACHE';
exports.DID_POOL_CACHE_LIMIT = 500;
let IndyPoolService = class IndyPoolService {
    constructor(cacheRepository, agentDependencies, logger, stop$, fileSystem) {
        this.pools = [];
        this.logger = logger;
        this.indy = agentDependencies.indy;
        this.agentDependencies = agentDependencies;
        this.fileSystem = fileSystem;
        this.stop$ = stop$;
        this.didCache = new cache_1.PersistedLruCache(exports.DID_POOL_CACHE_ID, exports.DID_POOL_CACHE_LIMIT, cacheRepository);
    }
    setPools(poolConfigs) {
        this.pools = poolConfigs.map((poolConfig) => new IndyPool_1.IndyPool(poolConfig, this.agentDependencies, this.logger, this.stop$, this.fileSystem));
    }
    /**
     * Create connections to all ledger pools
     */
    async connectToPools() {
        const handleArray = [];
        // Sequentially connect to pools so we don't use up too many resources connecting in parallel
        for (const pool of this.pools) {
            this.logger.debug(`Connecting to pool: ${pool.id}`);
            const poolHandle = await pool.connect();
            this.logger.debug(`Finished connection to pool: ${pool.id}`);
            handleArray.push(poolHandle);
        }
        return handleArray;
    }
    /**
     * @deprecated use instead getPoolForNamespace
     * Get the pool used for writing to the ledger. For now we always use the first pool
     *  as the pool that writes to the ledger
     */
    get ledgerWritePool() {
        if (this.pools.length === 0) {
            throw new LedgerNotConfiguredError_1.LedgerNotConfiguredError("No indy ledgers configured. Provide at least one pool configuration in the 'indyLedgers' agent configuration");
        }
        return this.pools[0];
    }
    /**
     * Get the most appropriate pool for the given did. The algorithm is based on the approach as described in this document:
     * https://docs.google.com/document/d/109C_eMsuZnTnYe2OAd02jAts1vC4axwEKIq7_4dnNVA/edit
     */
    async getPoolForDid(agentContext, did) {
        var _a;
        const pools = this.pools;
        if (pools.length === 0) {
            throw new LedgerNotConfiguredError_1.LedgerNotConfiguredError("No indy ledgers configured. Provide at least one pool configuration in the 'indyLedgers' agent configuration");
        }
        const cachedNymResponse = await this.didCache.get(agentContext, did);
        const pool = this.pools.find((pool) => pool.id === (cachedNymResponse === null || cachedNymResponse === void 0 ? void 0 : cachedNymResponse.poolId));
        // If we have the nym response with associated pool in the cache, we'll use that
        if (cachedNymResponse && pool) {
            this.logger.trace(`Found ledger id '${pool.id}' for did '${did}' in cache`);
            return { did: cachedNymResponse.nymResponse, pool };
        }
        const { successful, rejected } = await this.getSettledDidResponsesFromPools(did, pools);
        if (successful.length === 0) {
            const allNotFound = rejected.every((e) => e.reason instanceof LedgerNotFoundError_1.LedgerNotFoundError);
            const rejectedOtherThanNotFound = rejected.filter((e) => !(e.reason instanceof LedgerNotFoundError_1.LedgerNotFoundError));
            // All ledgers returned response that the did was not found
            if (allNotFound) {
                throw new LedgerNotFoundError_1.LedgerNotFoundError(`Did '${did}' not found on any of the ledgers (total ${this.pools.length}).`);
            }
            // one or more of the ledgers returned an unknown error
            throw new LedgerError_1.LedgerError(`Unknown error retrieving did '${did}' from '${rejectedOtherThanNotFound.length}' of '${pools.length}' ledgers`, { cause: rejectedOtherThanNotFound[0].reason });
        }
        // If there are self certified DIDs we always prefer it over non self certified DIDs
        // We take the first self certifying DID as we take the order in the
        // indyLedgers config as the order of preference of ledgers
        let value = (_a = successful.find((response) => (0, did_1.isSelfCertifiedDid)(response.value.did.did, response.value.did.verkey))) === null || _a === void 0 ? void 0 : _a.value;
        if (!value) {
            // Split between production and nonProduction ledgers. If there is at least one
            // successful response from a production ledger, only keep production ledgers
            // otherwise we only keep the non production ledgers.
            const production = successful.filter((s) => s.value.pool.config.isProduction);
            const nonProduction = successful.filter((s) => !s.value.pool.config.isProduction);
            const productionOrNonProduction = production.length >= 1 ? production : nonProduction;
            // We take the first value as we take the order in the indyLedgers config as
            // the order of preference of ledgers
            value = productionOrNonProduction[0].value;
        }
        await this.didCache.set(agentContext, did, {
            nymResponse: value.did,
            poolId: value.pool.id,
        });
        return { pool: value.pool, did: value.did };
    }
    async getSettledDidResponsesFromPools(did, pools) {
        this.logger.trace(`Retrieving did '${did}' from ${pools.length} ledgers`);
        const didResponses = await (0, promises_1.allSettled)(pools.map((pool) => this.getDidFromPool(did, pool)));
        const successful = (0, promises_1.onlyFulfilled)(didResponses);
        this.logger.trace(`Retrieved ${successful.length} responses from ledgers for did '${did}'`);
        const rejected = (0, promises_1.onlyRejected)(didResponses);
        return {
            rejected,
            successful,
        };
    }
    /**
     * Get the most appropriate pool for the given indyNamespace
     */
    getPoolForNamespace(indyNamespace) {
        if (this.pools.length === 0) {
            throw new LedgerNotConfiguredError_1.LedgerNotConfiguredError("No indy ledgers configured. Provide at least one pool configuration in the 'indyLedgers' agent configuration");
        }
        if (!indyNamespace) {
            this.logger.warn('Not passing the indyNamespace is deprecated and will be removed in the future version.');
            return this.pools[0];
        }
        const pool = this.pools.find((pool) => pool.didIndyNamespace === indyNamespace);
        if (!pool) {
            throw new LedgerNotFoundError_1.LedgerNotFoundError(`No ledgers found for IndyNamespace '${indyNamespace}'.`);
        }
        return pool;
    }
    async submitWriteRequest(agentContext, pool, request, signDid) {
        try {
            const requestWithTaa = await this.appendTaa(pool, request);
            const signedRequestWithTaa = await this.signRequest(agentContext, signDid, requestWithTaa);
            const response = await pool.submitWriteRequest(signedRequestWithTaa);
            return response;
        }
        catch (error) {
            throw (0, indyError_1.isIndyError)(error) ? new IndySdkError_1.IndySdkError(error) : error;
        }
    }
    async submitReadRequest(pool, request) {
        try {
            const response = await pool.submitReadRequest(request);
            return response;
        }
        catch (error) {
            throw (0, indyError_1.isIndyError)(error) ? new IndySdkError_1.IndySdkError(error) : error;
        }
    }
    async signRequest(agentContext, did, request) {
        (0, assertIndyWallet_1.assertIndyWallet)(agentContext.wallet);
        try {
            return this.indy.signRequest(agentContext.wallet.handle, did, request);
        }
        catch (error) {
            throw (0, indyError_1.isIndyError)(error) ? new IndySdkError_1.IndySdkError(error) : error;
        }
    }
    async appendTaa(pool, request) {
        try {
            const authorAgreement = await this.getTransactionAuthorAgreement(pool);
            const taa = pool.config.transactionAuthorAgreement;
            // If ledger does not have TAA, we can just send request
            if (authorAgreement == null) {
                return request;
            }
            // Ledger has taa but user has not specified which one to use
            if (!taa) {
                throw new LedgerError_1.LedgerError(`Please, specify a transaction author agreement with version and acceptance mechanism. ${JSON.stringify(authorAgreement)}`);
            }
            // Throw an error if the pool doesn't have the specified version and acceptance mechanism
            if (authorAgreement.version !== taa.version ||
                !(taa.acceptanceMechanism in authorAgreement.acceptanceMechanisms.aml)) {
                // Throw an error with a helpful message
                const errMessage = `Unable to satisfy matching TAA with mechanism ${JSON.stringify(taa.acceptanceMechanism)} and version ${JSON.stringify(taa.version)} in pool.\n Found ${JSON.stringify(Object.keys(authorAgreement.acceptanceMechanisms.aml))} and version ${authorAgreement.version} in pool.`;
                throw new LedgerError_1.LedgerError(errMessage);
            }
            const requestWithTaa = await this.indy.appendTxnAuthorAgreementAcceptanceToRequest(request, authorAgreement.text, taa.version, authorAgreement.digest, taa.acceptanceMechanism, 
            // Current time since epoch
            // We can't use ratification_ts, as it must be greater than 1499906902
            Math.floor(new Date().getTime() / 1000));
            return requestWithTaa;
        }
        catch (error) {
            throw (0, indyError_1.isIndyError)(error) ? new IndySdkError_1.IndySdkError(error) : error;
        }
    }
    async getTransactionAuthorAgreement(pool) {
        try {
            // TODO Replace this condition with memoization
            if (pool.authorAgreement !== undefined) {
                return pool.authorAgreement;
            }
            const taaRequest = await this.indy.buildGetTxnAuthorAgreementRequest(null);
            const taaResponse = await this.submitReadRequest(pool, taaRequest);
            const acceptanceMechanismRequest = await this.indy.buildGetAcceptanceMechanismsRequest(null);
            const acceptanceMechanismResponse = await this.submitReadRequest(pool, acceptanceMechanismRequest);
            // TAA can be null
            if (taaResponse.result.data == null) {
                pool.authorAgreement = null;
                return null;
            }
            // If TAA is not null, we can be sure AcceptanceMechanisms is also not null
            const authorAgreement = taaResponse.result.data;
            const acceptanceMechanisms = acceptanceMechanismResponse.result.data;
            pool.authorAgreement = Object.assign(Object.assign({}, authorAgreement), { acceptanceMechanisms });
            return pool.authorAgreement;
        }
        catch (error) {
            throw (0, indyError_1.isIndyError)(error) ? new IndySdkError_1.IndySdkError(error) : error;
        }
    }
    async getDidFromPool(did, pool) {
        try {
            this.logger.trace(`Get public did '${did}' from ledger '${pool.id}'`);
            const request = await this.indy.buildGetNymRequest(null, did);
            this.logger.trace(`Submitting get did request for did '${did}' to ledger '${pool.id}'`);
            const response = await pool.submitReadRequest(request);
            const result = await this.indy.parseGetNymResponse(response);
            this.logger.trace(`Retrieved did '${did}' from ledger '${pool.id}'`, result);
            return {
                did: result,
                pool,
                response,
            };
        }
        catch (error) {
            this.logger.trace(`Error retrieving did '${did}' from ledger '${pool.id}'`, {
                error,
                did,
            });
            if ((0, indyError_1.isIndyError)(error, 'LedgerNotFound')) {
                throw new LedgerNotFoundError_1.LedgerNotFoundError(`Did '${did}' not found on ledger ${pool.id}`);
            }
            else {
                throw (0, indyError_1.isIndyError)(error) ? new IndySdkError_1.IndySdkError(error) : error;
            }
        }
    }
};
IndyPoolService = __decorate([
    (0, plugins_1.injectable)(),
    __param(1, (0, plugins_1.inject)(constants_1.InjectionSymbols.AgentDependencies)),
    __param(2, (0, plugins_1.inject)(constants_1.InjectionSymbols.Logger)),
    __param(3, (0, plugins_1.inject)(constants_1.InjectionSymbols.Stop$)),
    __param(4, (0, plugins_1.inject)(constants_1.InjectionSymbols.FileSystem)),
    __metadata("design:paramtypes", [cache_1.CacheRepository, Object, Object, rxjs_1.Subject, Object])
], IndyPoolService);
exports.IndyPoolService = IndyPoolService;
