// @aries-framework/core/build/modules/ledger/services/IndyLedgerService.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.IndyLedgerService = void 0;
const constants_1 = require("../../../constants");
const IndySdkError_1 = require("../../../error/IndySdkError");
const plugins_1 = require("../../../plugins");
const did_1 = require("../../../utils/did");
const indyError_1 = require("../../../utils/indyError");
const IndyIssuerService_1 = require("../../indy/services/IndyIssuerService");
const IndyPoolService_1 = require("./IndyPoolService");
let IndyLedgerService = class IndyLedgerService {
    constructor(agentDependencies, logger, indyIssuer, indyPoolService) {
        this.indy = agentDependencies.indy;
        this.logger = logger;
        this.indyIssuer = indyIssuer;
        this.indyPoolService = indyPoolService;
    }
    setPools(poolConfigs) {
        return this.indyPoolService.setPools(poolConfigs);
    }
    /**
     * @deprecated
     */
    getDidIndyWriteNamespace() {
        return this.indyPoolService.ledgerWritePool.config.indyNamespace;
    }
    async connectToPools() {
        return this.indyPoolService.connectToPools();
    }
    /**
     * @deprecated
     */
    async registerPublicDid(agentContext, submitterDid, targetDid, verkey, alias, role) {
        const pool = this.indyPoolService.getPoolForNamespace();
        try {
            this.logger.debug(`Register public did '${targetDid}' on ledger '${pool.id}'`);
            const request = await this.indy.buildNymRequest(submitterDid, targetDid, verkey, alias, role || null);
            const response = await this.indyPoolService.submitWriteRequest(agentContext, pool, request, submitterDid);
            this.logger.debug(`Registered public did '${targetDid}' on ledger '${pool.id}'`, {
                response,
            });
            return targetDid;
        }
        catch (error) {
            this.logger.error(`Error registering public did '${targetDid}' on ledger '${pool.id}'`, {
                error,
                submitterDid,
                targetDid,
                verkey,
                alias,
                role,
                pool: pool.id,
            });
            throw error;
        }
    }
    /**
     * @deprecated
     */
    async getPublicDid(agentContext, did) {
        // Getting the pool for a did also retrieves the DID. We can just use that
        const { did: didResponse } = await this.indyPoolService.getPoolForDid(agentContext, did);
        return didResponse;
    }
    /**
     * @deprecated
     */
    async setEndpointsForDid(agentContext, did, endpoints) {
        const pool = this.indyPoolService.getPoolForNamespace();
        try {
            this.logger.debug(`Set endpoints for did '${did}' on ledger '${pool.id}'`, endpoints);
            const request = await this.indy.buildAttribRequest(did, did, null, { endpoint: endpoints }, null);
            const response = await this.indyPoolService.submitWriteRequest(agentContext, pool, request, did);
            this.logger.debug(`Successfully set endpoints for did '${did}' on ledger '${pool.id}'`, {
                response,
                endpoints,
            });
        }
        catch (error) {
            this.logger.error(`Error setting endpoints for did '${did}' on ledger '${pool.id}'`, {
                error,
                did,
                endpoints,
            });
            throw (0, indyError_1.isIndyError)(error) ? new IndySdkError_1.IndySdkError(error) : error;
        }
    }
    /**
     * @deprecated
     */
    async getEndpointsForDid(agentContext, did) {
        var _a;
        const { pool } = await this.indyPoolService.getPoolForDid(agentContext, did);
        try {
            this.logger.debug(`Get endpoints for did '${did}' from ledger '${pool.id}'`);
            const request = await this.indy.buildGetAttribRequest(null, did, 'endpoint', null, null);
            this.logger.debug(`Submitting get endpoint ATTRIB request for did '${did}' to ledger '${pool.id}'`);
            const response = await this.indyPoolService.submitReadRequest(pool, request);
            if (!response.result.data)
                return {};
            const endpoints = (_a = JSON.parse(response.result.data)) === null || _a === void 0 ? void 0 : _a.endpoint;
            this.logger.debug(`Got endpoints '${JSON.stringify(endpoints)}' for did '${did}' from ledger '${pool.id}'`, {
                response,
                endpoints,
            });
            return endpoints !== null && endpoints !== void 0 ? endpoints : {};
        }
        catch (error) {
            this.logger.error(`Error retrieving endpoints for did '${did}' from ledger '${pool.id}'`, {
                error,
            });
            throw (0, indyError_1.isIndyError)(error) ? new IndySdkError_1.IndySdkError(error) : error;
        }
    }
    async registerSchema(agentContext, did, schemaTemplate) {
        const pool = this.indyPoolService.getPoolForNamespace();
        try {
            this.logger.debug(`Register schema on ledger '${pool.id}' with did '${did}'`, schemaTemplate);
            const { name, attributes, version } = schemaTemplate;
            const schema = await this.indyIssuer.createSchema(agentContext, { originDid: did, name, version, attributes });
            const request = await this.indy.buildSchemaRequest(did, schema);
            const response = await this.indyPoolService.submitWriteRequest(agentContext, pool, request, did);
            this.logger.debug(`Registered schema '${schema.id}' on ledger '${pool.id}'`, {
                response,
                schema,
            });
            schema.seqNo = response.result.txnMetadata.seqNo;
            return schema;
        }
        catch (error) {
            this.logger.error(`Error registering schema for did '${did}' on ledger '${pool.id}'`, {
                error,
                did,
                schemaTemplate,
            });
            throw (0, indyError_1.isIndyError)(error) ? new IndySdkError_1.IndySdkError(error) : error;
        }
    }
    async getSchema(agentContext, schemaId) {
        const did = (0, did_1.didFromSchemaId)(schemaId);
        const { pool } = await this.indyPoolService.getPoolForDid(agentContext, did);
        try {
            this.logger.debug(`Getting schema '${schemaId}' from ledger '${pool.id}'`);
            const request = await this.indy.buildGetSchemaRequest(null, schemaId);
            this.logger.trace(`Submitting get schema request for schema '${schemaId}' to ledger '${pool.id}'`);
            const response = await this.indyPoolService.submitReadRequest(pool, request);
            this.logger.trace(`Got un-parsed schema '${schemaId}' from ledger '${pool.id}'`, {
                response,
            });
            const [, schema] = await this.indy.parseGetSchemaResponse(response);
            this.logger.debug(`Got schema '${schemaId}' from ledger '${pool.id}'`, {
                schema,
            });
            return schema;
        }
        catch (error) {
            this.logger.error(`Error retrieving schema '${schemaId}' from ledger '${pool.id}'`, {
                error,
                schemaId,
            });
            throw (0, indyError_1.isIndyError)(error) ? new IndySdkError_1.IndySdkError(error) : error;
        }
    }
    async registerCredentialDefinition(agentContext, did, credentialDefinitionTemplate) {
        const pool = this.indyPoolService.getPoolForNamespace();
        try {
            this.logger.debug(`Registering credential definition on ledger '${pool.id}' with did '${did}'`, credentialDefinitionTemplate);
            const { schema, tag, signatureType, supportRevocation } = credentialDefinitionTemplate;
            const credentialDefinition = await this.indyIssuer.createCredentialDefinition(agentContext, {
                issuerDid: did,
                schema,
                tag,
                signatureType,
                supportRevocation,
            });
            const request = await this.indy.buildCredDefRequest(did, credentialDefinition);
            const response = await this.indyPoolService.submitWriteRequest(agentContext, pool, request, did);
            this.logger.debug(`Registered credential definition '${credentialDefinition.id}' on ledger '${pool.id}'`, {
                response,
                credentialDefinition: credentialDefinition,
            });
            return credentialDefinition;
        }
        catch (error) {
            this.logger.error(`Error registering credential definition for schema '${credentialDefinitionTemplate.schema.id}' on ledger '${pool.id}'`, {
                error,
                did,
                credentialDefinitionTemplate,
            });
            throw (0, indyError_1.isIndyError)(error) ? new IndySdkError_1.IndySdkError(error) : error;
        }
    }
    async getCredentialDefinition(agentContext, credentialDefinitionId) {
        const did = (0, did_1.didFromCredentialDefinitionId)(credentialDefinitionId);
        const { pool } = await this.indyPoolService.getPoolForDid(agentContext, did);
        this.logger.debug(`Using ledger '${pool.id}' to retrieve credential definition '${credentialDefinitionId}'`);
        try {
            const request = await this.indy.buildGetCredDefRequest(null, credentialDefinitionId);
            this.logger.trace(`Submitting get credential definition request for credential definition '${credentialDefinitionId}' to ledger '${pool.id}'`);
            const response = await this.indyPoolService.submitReadRequest(pool, request);
            this.logger.trace(`Got un-parsed credential definition '${credentialDefinitionId}' from ledger '${pool.id}'`, {
                response,
            });
            const [, credentialDefinition] = await this.indy.parseGetCredDefResponse(response);
            this.logger.debug(`Got credential definition '${credentialDefinitionId}' from ledger '${pool.id}'`, {
                credentialDefinition,
            });
            return credentialDefinition;
        }
        catch (error) {
            this.logger.error(`Error retrieving credential definition '${credentialDefinitionId}' from ledger '${pool.id}'`, {
                error,
                credentialDefinitionId,
                pool: pool.id,
            });
            throw (0, indyError_1.isIndyError)(error) ? new IndySdkError_1.IndySdkError(error) : error;
        }
    }
    async getRevocationRegistryDefinition(agentContext, revocationRegistryDefinitionId) {
        const did = (0, did_1.didFromRevocationRegistryDefinitionId)(revocationRegistryDefinitionId);
        const { pool } = await this.indyPoolService.getPoolForDid(agentContext, did);
        this.logger.debug(`Using ledger '${pool.id}' to retrieve revocation registry definition '${revocationRegistryDefinitionId}'`);
        try {
            //TODO - implement a cache
            this.logger.trace(`Revocation Registry Definition '${revocationRegistryDefinitionId}' not cached, retrieving from ledger`);
            const request = await this.indy.buildGetRevocRegDefRequest(null, revocationRegistryDefinitionId);
            this.logger.trace(`Submitting get revocation registry definition request for revocation registry definition '${revocationRegistryDefinitionId}' to ledger`);
            const response = await this.indyPoolService.submitReadRequest(pool, request);
            this.logger.trace(`Got un-parsed revocation registry definition '${revocationRegistryDefinitionId}' from ledger '${pool.id}'`, {
                response,
            });
            const [, revocationRegistryDefinition] = await this.indy.parseGetRevocRegDefResponse(response);
            this.logger.debug(`Got revocation registry definition '${revocationRegistryDefinitionId}' from ledger`, {
                revocationRegistryDefinition,
            });
            return { revocationRegistryDefinition, revocationRegistryDefinitionTxnTime: response.result.txnTime };
        }
        catch (error) {
            this.logger.error(`Error retrieving revocation registry definition '${revocationRegistryDefinitionId}' from ledger`, {
                error,
                revocationRegistryDefinitionId: revocationRegistryDefinitionId,
                pool: pool.id,
            });
            throw error;
        }
    }
    // Retrieves the accumulated state of a revocation registry by id given a revocation interval from & to (used primarily for proof creation)
    async getRevocationRegistryDelta(agentContext, revocationRegistryDefinitionId, to = new Date().getTime(), from = 0) {
        //TODO - implement a cache
        const did = (0, did_1.didFromRevocationRegistryDefinitionId)(revocationRegistryDefinitionId);
        const { pool } = await this.indyPoolService.getPoolForDid(agentContext, did);
        this.logger.debug(`Using ledger '${pool.id}' to retrieve revocation registry delta with revocation registry definition id: '${revocationRegistryDefinitionId}'`, {
            to,
            from,
        });
        try {
            const request = await this.indy.buildGetRevocRegDeltaRequest(null, revocationRegistryDefinitionId, from, to);
            this.logger.trace(`Submitting get revocation registry delta request for revocation registry '${revocationRegistryDefinitionId}' to ledger`);
            const response = await this.indyPoolService.submitReadRequest(pool, request);
            this.logger.trace(`Got revocation registry delta unparsed-response '${revocationRegistryDefinitionId}' from ledger`, {
                response,
            });
            const [, revocationRegistryDelta, deltaTimestamp] = await this.indy.parseGetRevocRegDeltaResponse(response);
            this.logger.debug(`Got revocation registry delta '${revocationRegistryDefinitionId}' from ledger`, {
                revocationRegistryDelta,
                deltaTimestamp,
                to,
                from,
            });
            return { revocationRegistryDelta, deltaTimestamp };
        }
        catch (error) {
            this.logger.error(`Error retrieving revocation registry delta '${revocationRegistryDefinitionId}' from ledger, potentially revocation interval ends before revocation registry creation?"`, {
                error,
                revocationRegistryId: revocationRegistryDefinitionId,
                pool: pool.id,
            });
            throw error;
        }
    }
    // Retrieves the accumulated state of a revocation registry by id given a timestamp (used primarily for verification)
    async getRevocationRegistry(agentContext, revocationRegistryDefinitionId, timestamp) {
        //TODO - implement a cache
        const did = (0, did_1.didFromRevocationRegistryDefinitionId)(revocationRegistryDefinitionId);
        const { pool } = await this.indyPoolService.getPoolForDid(agentContext, did);
        this.logger.debug(`Using ledger '${pool.id}' to retrieve revocation registry accumulated state with revocation registry definition id: '${revocationRegistryDefinitionId}'`, {
            timestamp,
        });
        try {
            const request = await this.indy.buildGetRevocRegRequest(null, revocationRegistryDefinitionId, timestamp);
            this.logger.trace(`Submitting get revocation registry request for revocation registry '${revocationRegistryDefinitionId}' to ledger`);
            const response = await this.indyPoolService.submitReadRequest(pool, request);
            this.logger.trace(`Got un-parsed revocation registry '${revocationRegistryDefinitionId}' from ledger '${pool.id}'`, {
                response,
            });
            const [, revocationRegistry, ledgerTimestamp] = await this.indy.parseGetRevocRegResponse(response);
            this.logger.debug(`Got revocation registry '${revocationRegistryDefinitionId}' from ledger`, {
                ledgerTimestamp,
                revocationRegistry,
            });
            return { revocationRegistry, ledgerTimestamp };
        }
        catch (error) {
            this.logger.error(`Error retrieving revocation registry '${revocationRegistryDefinitionId}' from ledger`, {
                error,
                revocationRegistryId: revocationRegistryDefinitionId,
                pool: pool.id,
            });
            throw error;
        }
    }
};
IndyLedgerService = __decorate([
    (0, plugins_1.injectable)(),
    __param(0, (0, plugins_1.inject)(constants_1.InjectionSymbols.AgentDependencies)),
    __param(1, (0, plugins_1.inject)(constants_1.InjectionSymbols.Logger)),
    __metadata("design:paramtypes", [Object, Object, IndyIssuerService_1.IndyIssuerService,
        IndyPoolService_1.IndyPoolService])
], IndyLedgerService);
exports.IndyLedgerService = IndyLedgerService;
