// @aries-framework/core/build/modules/dids/methods/sov/IndySdkSovDidResolver.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.IndySdkSovDidResolver = void 0;
const error_1 = require("../../../../error");
const plugins_1 = require("../../../../plugins");
const indyError_1 = require("../../../../utils/indyError");
const ledger_1 = require("../../../ledger");
const util_1 = require("./util");
let IndySdkSovDidResolver = class IndySdkSovDidResolver {
    constructor() {
        this.supportedMethods = ['sov'];
    }
    async resolve(agentContext, did, parsed) {
        const didDocumentMetadata = {};
        try {
            const nym = await this.getPublicDid(agentContext, parsed.id);
            const endpoints = await this.getEndpointsForDid(agentContext, parsed.id);
            const keyAgreementId = `${parsed.did}#key-agreement-1`;
            const builder = (0, util_1.sovDidDocumentFromDid)(parsed.did, nym.verkey);
            (0, util_1.addServicesFromEndpointsAttrib)(builder, parsed.did, endpoints, keyAgreementId);
            return {
                didDocument: builder.build(),
                didDocumentMetadata,
                didResolutionMetadata: { contentType: 'application/did+ld+json' },
            };
        }
        catch (error) {
            return {
                didDocument: null,
                didDocumentMetadata,
                didResolutionMetadata: {
                    error: 'notFound',
                    message: `resolver_error: Unable to resolve did '${did}': ${error}`,
                },
            };
        }
    }
    async getPublicDid(agentContext, did) {
        const indyPoolService = agentContext.dependencyManager.resolve(ledger_1.IndyPoolService);
        // Getting the pool for a did also retrieves the DID. We can just use that
        const { did: didResponse } = await indyPoolService.getPoolForDid(agentContext, did);
        return didResponse;
    }
    async getEndpointsForDid(agentContext, did) {
        var _a;
        const indyPoolService = agentContext.dependencyManager.resolve(ledger_1.IndyPoolService);
        const indy = agentContext.config.agentDependencies.indy;
        const { pool } = await indyPoolService.getPoolForDid(agentContext, did);
        try {
            agentContext.config.logger.debug(`Get endpoints for did '${did}' from ledger '${pool.id}'`);
            const request = await indy.buildGetAttribRequest(null, did, 'endpoint', null, null);
            agentContext.config.logger.debug(`Submitting get endpoint ATTRIB request for did '${did}' to ledger '${pool.id}'`);
            const response = await indyPoolService.submitReadRequest(pool, request);
            if (!response.result.data)
                return {};
            const endpoints = (_a = JSON.parse(response.result.data)) === null || _a === void 0 ? void 0 : _a.endpoint;
            agentContext.config.logger.debug(`Got endpoints '${JSON.stringify(endpoints)}' for did '${did}' from ledger '${pool.id}'`, {
                response,
                endpoints,
            });
            return endpoints !== null && endpoints !== void 0 ? endpoints : {};
        }
        catch (error) {
            agentContext.config.logger.error(`Error retrieving endpoints for did '${did}' from ledger '${pool.id}'`, {
                error,
            });
            throw (0, indyError_1.isIndyError)(error) ? new error_1.IndySdkError(error) : error;
        }
    }
};
IndySdkSovDidResolver = __decorate([
    (0, plugins_1.injectable)()
], IndySdkSovDidResolver);
exports.IndySdkSovDidResolver = IndySdkSovDidResolver;
