// @aries-framework/core/build/modules/vc/repository/W3cCredentialRecord.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.W3cCredentialRecord = void 0;
const class_transformer_1 = require("class-transformer");
const BaseRecord_1 = require("../../../storage/BaseRecord");
const uuid_1 = require("../../../utils/uuid");
const W3cVerifiableCredential_1 = require("../models/credential/W3cVerifiableCredential");
class W3cCredentialRecord extends BaseRecord_1.BaseRecord {
    constructor(props) {
        var _a, _b;
        super();
        this.type = W3cCredentialRecord.type;
        if (props) {
            this.id = (_a = props.id) !== null && _a !== void 0 ? _a : (0, uuid_1.uuid)();
            this.createdAt = (_b = props.createdAt) !== null && _b !== void 0 ? _b : new Date();
            this._tags = props.tags;
            this.credential = props.credential;
        }
    }
    getTags() {
        return Object.assign(Object.assign({}, this._tags), { issuerId: this.credential.issuerId, subjectIds: this.credential.credentialSubjectIds, schemaIds: this.credential.credentialSchemaIds, contexts: this.credential.contexts, proofTypes: this.credential.proofTypes, givenId: this.credential.id });
    }
}
W3cCredentialRecord.type = 'W3cCredentialRecord';
__decorate([
    (0, class_transformer_1.Type)(() => W3cVerifiableCredential_1.W3cVerifiableCredential),
    __metadata("design:type", W3cVerifiableCredential_1.W3cVerifiableCredential)
], W3cCredentialRecord.prototype, "credential", void 0);
exports.W3cCredentialRecord = W3cCredentialRecord;
