// @aries-framework/core/build/modules/routing/RecipientModule.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RecipientModule = void 0;
const models_1 = require("../../agent/models");
const RecipientApi_1 = require("./RecipientApi");
const RecipientModuleConfig_1 = require("./RecipientModuleConfig");
const models_2 = require("./models");
const repository_1 = require("./repository");
const services_1 = require("./services");
class RecipientModule {
    constructor(config) {
        this.api = RecipientApi_1.RecipientApi;
        this.config = new RecipientModuleConfig_1.RecipientModuleConfig(config);
    }
    /**
     * Registers the dependencies of the mediator recipient module on the dependency manager.
     */
    register(dependencyManager, featureRegistry) {
        // Api
        dependencyManager.registerContextScoped(RecipientApi_1.RecipientApi);
        // Config
        dependencyManager.registerInstance(RecipientModuleConfig_1.RecipientModuleConfig, this.config);
        // Services
        dependencyManager.registerSingleton(services_1.MediationRecipientService);
        dependencyManager.registerSingleton(services_1.RoutingService);
        // Repositories
        dependencyManager.registerSingleton(repository_1.MediationRepository);
        // Features
        featureRegistry.register(new models_1.Protocol({
            id: 'https://didcomm.org/coordinate-mediation/1.0',
            roles: [models_2.MediationRole.Recipient],
        }));
    }
}
exports.RecipientModule = RecipientModule;
