// @aries-framework/core/build/modules/proofs/ProofResponseCoordinator.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var ProofResponseCoordinator_1;
Object.defineProperty(exports, "__esModule", { value: true });
exports.ProofResponseCoordinator = void 0;
const plugins_1 = require("../../plugins");
const ProofService_1 = require("./ProofService");
const ProofAutoAcceptType_1 = require("./models/ProofAutoAcceptType");
/**
 * This class handles all the automation with all the messages in the present proof protocol
 * Every function returns `true` if it should automate the flow and `false` if not
 */
let ProofResponseCoordinator = ProofResponseCoordinator_1 = class ProofResponseCoordinator {
    constructor(proofService) {
        this.proofService = proofService;
    }
    /**
     * Returns the proof auto accept config based on priority:
     *	- The record config takes first priority
     *	- Otherwise the agent config
     *	- Otherwise {@link AutoAcceptProof.Never} is returned
     */
    static composeAutoAccept(recordConfig, agentConfig) {
        var _a;
        return (_a = recordConfig !== null && recordConfig !== void 0 ? recordConfig : agentConfig) !== null && _a !== void 0 ? _a : ProofAutoAcceptType_1.AutoAcceptProof.Never;
    }
    /**
     * Checks whether it should automatically respond to a proposal
     */
    async shouldAutoRespondToProposal(agentContext, proofRecord) {
        const autoAccept = ProofResponseCoordinator_1.composeAutoAccept(proofRecord.autoAcceptProof, agentContext.config.autoAcceptProofs);
        if (autoAccept === ProofAutoAcceptType_1.AutoAcceptProof.Always) {
            return true;
        }
        if (autoAccept === ProofAutoAcceptType_1.AutoAcceptProof.ContentApproved) {
            return this.proofService.shouldAutoRespondToProposal(agentContext, proofRecord);
        }
        return false;
    }
    /**
     * Checks whether it should automatically respond to a request
     */
    async shouldAutoRespondToRequest(agentContext, proofRecord) {
        const autoAccept = ProofResponseCoordinator_1.composeAutoAccept(proofRecord.autoAcceptProof, agentContext.config.autoAcceptProofs);
        if (autoAccept === ProofAutoAcceptType_1.AutoAcceptProof.Always) {
            return true;
        }
        if (autoAccept === ProofAutoAcceptType_1.AutoAcceptProof.ContentApproved) {
            return this.proofService.shouldAutoRespondToRequest(agentContext, proofRecord);
        }
        return false;
    }
    /**
     * Checks whether it should automatically respond to a presentation of proof
     */
    async shouldAutoRespondToPresentation(agentContext, proofRecord) {
        const autoAccept = ProofResponseCoordinator_1.composeAutoAccept(proofRecord.autoAcceptProof, agentContext.config.autoAcceptProofs);
        if (autoAccept === ProofAutoAcceptType_1.AutoAcceptProof.Always) {
            return true;
        }
        if (autoAccept === ProofAutoAcceptType_1.AutoAcceptProof.ContentApproved) {
            return this.proofService.shouldAutoRespondToPresentation(agentContext, proofRecord);
        }
        return false;
    }
};
ProofResponseCoordinator = ProofResponseCoordinator_1 = __decorate([
    (0, plugins_1.injectable)(),
    __metadata("design:paramtypes", [ProofService_1.ProofService])
], ProofResponseCoordinator);
exports.ProofResponseCoordinator = ProofResponseCoordinator;
