// @aries-framework/core/build/modules/proofs/ProofsModule.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ProofsModule = void 0;
const models_1 = require("../../agent/models");
const ProofsApi_1 = require("./ProofsApi");
const ProofsModuleConfig_1 = require("./ProofsModuleConfig");
const IndyProofFormatService_1 = require("./formats/indy/IndyProofFormatService");
const v1_1 = require("./protocol/v1");
const v2_1 = require("./protocol/v2");
const repository_1 = require("./repository");
class ProofsModule {
    constructor(config) {
        this.api = ProofsApi_1.ProofsApi;
        this.config = new ProofsModuleConfig_1.ProofsModuleConfig(config);
    }
    /**
     * Registers the dependencies of the proofs module on the dependency manager.
     */
    register(dependencyManager, featureRegistry) {
        // Api
        dependencyManager.registerContextScoped(ProofsApi_1.ProofsApi);
        // Config
        dependencyManager.registerInstance(ProofsModuleConfig_1.ProofsModuleConfig, this.config);
        // Services
        dependencyManager.registerSingleton(v1_1.V1ProofService);
        dependencyManager.registerSingleton(v2_1.V2ProofService);
        // Repositories
        dependencyManager.registerSingleton(repository_1.ProofRepository);
        // Proof Formats
        dependencyManager.registerSingleton(IndyProofFormatService_1.IndyProofFormatService);
        // Features
        featureRegistry.register(new models_1.Protocol({
            id: 'https://didcomm.org/present-proof/1.0',
            roles: ['verifier', 'prover'],
        }));
    }
}
exports.ProofsModule = ProofsModule;
