// @aries-framework/core/build/modules/proofs/ProofsApi.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ProofsApi = void 0;
const tsyringe_1 = require("tsyringe");
const AgentConfig_1 = require("../../agent/AgentConfig");
const Dispatcher_1 = require("../../agent/Dispatcher");
const MessageSender_1 = require("../../agent/MessageSender");
const AgentContext_1 = require("../../agent/context/AgentContext");
const models_1 = require("../../agent/models");
const constants_1 = require("../../constants");
const ServiceDecorator_1 = require("../../decorators/service/ServiceDecorator");
const error_1 = require("../../error");
const DidCommMessageRole_1 = require("../../storage/didcomm/DidCommMessageRole");
const ConnectionService_1 = require("../connections/services/ConnectionService");
const MediationRecipientService_1 = require("../routing/services/MediationRecipientService");
const RoutingService_1 = require("../routing/services/RoutingService");
const ProofResponseCoordinator_1 = require("./ProofResponseCoordinator");
const ProofState_1 = require("./models/ProofState");
const V1ProofService_1 = require("./protocol/v1/V1ProofService");
const V2ProofService_1 = require("./protocol/v2/V2ProofService");
const ProofRepository_1 = require("./repository/ProofRepository");
let ProofsApi = class ProofsApi {
    constructor(dispatcher, mediationRecipientService, messageSender, connectionService, agentContext, agentConfig, routingService, logger, proofRepository, v1Service, v2Service) {
        this.messageSender = messageSender;
        this.connectionService = connectionService;
        this.proofRepository = proofRepository;
        this.agentContext = agentContext;
        this.agentConfig = agentConfig;
        this.routingService = routingService;
        this.logger = logger;
        // Dynamically build service map. This will be extracted once services are registered dynamically
        this.serviceMap = [v1Service, v2Service].reduce((serviceMap, service) => (Object.assign(Object.assign({}, serviceMap), { [service.version]: service })), {});
        this.logger.debug(`Initializing Proofs Module for agent ${this.agentContext.config.label}`);
        this.registerMessageHandlers(dispatcher, mediationRecipientService);
    }
    getService(protocolVersion) {
        if (!this.serviceMap[protocolVersion]) {
            throw new error_1.AriesFrameworkError(`No proof service registered for protocol version ${protocolVersion}`);
        }
        return this.serviceMap[protocolVersion];
    }
    /**
     * Initiate a new presentation exchange as prover by sending a presentation proposal message
     * to the connection with the specified connection id.
     *
     * @param options multiple properties like protocol version, connection id, proof format (indy/ presentation exchange)
     * to include in the message
     * @returns Proof record associated with the sent proposal message
     */
    async proposeProof(options) {
        const service = this.getService(options.protocolVersion);
        const { connectionId } = options;
        const connection = await this.connectionService.getById(this.agentContext, connectionId);
        // Assert
        connection.assertReady();
        const proposalOptions = {
            connectionRecord: connection,
            proofFormats: options.proofFormats,
            autoAcceptProof: options.autoAcceptProof,
            goalCode: options.goalCode,
            comment: options.comment,
            parentThreadId: options.parentThreadId,
        };
        const { message, proofRecord } = await service.createProposal(this.agentContext, proposalOptions);
        const outboundMessageContext = new models_1.OutboundMessageContext(message, {
            agentContext: this.agentContext,
            connection,
            associatedRecord: proofRecord,
        });
        await this.messageSender.sendMessage(outboundMessageContext);
        return proofRecord;
    }
    /**
     * Accept a presentation proposal as verifier (by sending a presentation request message) to the connection
     * associated with the proof record.
     *
     * @param options multiple properties like proof record id, additional configuration for creating the request
     * @returns Proof record associated with the presentation request
     */
    async acceptProposal(options) {
        var _a;
        const { proofRecordId } = options;
        const proofRecord = await this.getById(proofRecordId);
        const service = this.getService(proofRecord.protocolVersion);
        if (!proofRecord.connectionId) {
            throw new error_1.AriesFrameworkError(`No connectionId found for proof record '${proofRecord.id}'. Connection-less issuance does not support presentation proposal or negotiation.`);
        }
        const connection = await this.connectionService.getById(this.agentContext, proofRecord.connectionId);
        // Assert
        connection.assertReady();
        const proofRequestFromProposalOptions = {
            proofRecord,
        };
        const proofRequest = await service.createProofRequestFromProposal(this.agentContext, proofRequestFromProposalOptions);
        const requestOptions = {
            proofRecord: proofRecord,
            proofFormats: proofRequest.proofFormats,
            goalCode: options.goalCode,
            willConfirm: (_a = options.willConfirm) !== null && _a !== void 0 ? _a : true,
            comment: options.comment,
        };
        const { message } = await service.createRequestAsResponse(this.agentContext, requestOptions);
        const outboundMessageContext = new models_1.OutboundMessageContext(message, {
            agentContext: this.agentContext,
            connection,
            associatedRecord: proofRecord,
        });
        await this.messageSender.sendMessage(outboundMessageContext);
        return proofRecord;
    }
    /**
     * Answer with a new presentation request in response to received presentation proposal message
     * to the connection associated with the proof record.
     *
     * @param options multiple properties like proof record id, proof formats to accept requested credentials object
     * specifying which credentials to use for the proof
     * @returns Proof record associated with the sent request message
     */
    async negotiateProposal(options) {
        const { proofRecordId } = options;
        const proofRecord = await this.getById(proofRecordId);
        const service = this.getService(proofRecord.protocolVersion);
        if (!proofRecord.connectionId) {
            throw new error_1.AriesFrameworkError(`No connectionId found for proof record '${proofRecord.id}'. Connection-less issuance does not support negotiation.`);
        }
        const connection = await this.connectionService.getById(this.agentContext, proofRecord.connectionId);
        // Assert
        connection.assertReady();
        const requestOptions = {
            proofRecord,
            proofFormats: options.proofFormats,
            autoAcceptProof: options.autoAcceptProof,
            comment: options.comment,
        };
        const { message } = await service.createRequestAsResponse(this.agentContext, requestOptions);
        const outboundMessageContext = new models_1.OutboundMessageContext(message, {
            agentContext: this.agentContext,
            connection,
            associatedRecord: proofRecord,
        });
        await this.messageSender.sendMessage(outboundMessageContext);
        return proofRecord;
    }
    /**
     * Initiate a new presentation exchange as verifier by sending a presentation request message
     * to the connection with the specified connection id
     *
     * @param options multiple properties like connection id, protocol version, proof Formats to build the proof request
     * @returns Proof record associated with the sent request message
     */
    async requestProof(options) {
        const service = this.getService(options.protocolVersion);
        const connection = await this.connectionService.getById(this.agentContext, options.connectionId);
        // Assert
        connection.assertReady();
        const createProofRequest = {
            connectionRecord: connection,
            proofFormats: options.proofFormats,
            autoAcceptProof: options.autoAcceptProof,
            parentThreadId: options.parentThreadId,
            comment: options.comment,
        };
        const { message, proofRecord } = await service.createRequest(this.agentContext, createProofRequest);
        const outboundMessageContext = new models_1.OutboundMessageContext(message, {
            agentContext: this.agentContext,
            connection,
            associatedRecord: proofRecord,
        });
        await this.messageSender.sendMessage(outboundMessageContext);
        return proofRecord;
    }
    /**
     * Accept a presentation request as prover (by sending a presentation message) to the connection
     * associated with the proof record.
     *
     * @param options multiple properties like proof record id, proof formats to accept requested credentials object
     * specifying which credentials to use for the proof
     * @returns Proof record associated with the sent presentation message
     */
    async acceptRequest(options) {
        const { proofRecordId, proofFormats, comment } = options;
        const record = await this.getById(proofRecordId);
        const service = this.getService(record.protocolVersion);
        const presentationOptions = {
            proofFormats,
            proofRecord: record,
            comment,
        };
        const { message, proofRecord } = await service.createPresentation(this.agentContext, presentationOptions);
        const requestMessage = await service.findRequestMessage(this.agentContext, proofRecord.id);
        // Use connection if present
        if (proofRecord.connectionId) {
            const connection = await this.connectionService.getById(this.agentContext, proofRecord.connectionId);
            // Assert
            connection.assertReady();
            const outboundMessageContext = new models_1.OutboundMessageContext(message, {
                agentContext: this.agentContext,
                connection,
                associatedRecord: proofRecord,
            });
            await this.messageSender.sendMessage(outboundMessageContext);
            return proofRecord;
        }
        // Use ~service decorator otherwise
        else if (requestMessage === null || requestMessage === void 0 ? void 0 : requestMessage.service) {
            // Create ~service decorator
            const routing = await this.routingService.getRouting(this.agentContext);
            message.service = new ServiceDecorator_1.ServiceDecorator({
                serviceEndpoint: routing.endpoints[0],
                recipientKeys: [routing.recipientKey.publicKeyBase58],
                routingKeys: routing.routingKeys.map((key) => key.publicKeyBase58),
            });
            const recipientService = requestMessage.service;
            // Set and save ~service decorator to record (to remember our verkey)
            await service.saveOrUpdatePresentationMessage(this.agentContext, {
                proofRecord: proofRecord,
                message: message,
                role: DidCommMessageRole_1.DidCommMessageRole.Sender,
            });
            await this.messageSender.sendMessageToService(new models_1.OutboundMessageContext(message, {
                agentContext: this.agentContext,
                serviceParams: {
                    service: recipientService.resolvedDidCommService,
                    senderKey: message.service.resolvedDidCommService.recipientKeys[0],
                    returnRoute: true,
                },
            }));
            return proofRecord;
        }
        // Cannot send message without connectionId or ~service decorator
        else {
            throw new error_1.AriesFrameworkError(`Cannot accept presentation request without connectionId or ~service decorator on presentation request.`);
        }
    }
    /**
     * Initiate a new presentation exchange as verifier by sending an out of band presentation
     * request message
     *
     * @param options multiple properties like protocol version, proof Formats to build the proof request
     * @returns the message itself and the proof record associated with the sent request message
     */
    async createRequest(options) {
        const service = this.getService(options.protocolVersion);
        const createProofRequest = {
            proofFormats: options.proofFormats,
            autoAcceptProof: options.autoAcceptProof,
            comment: options.comment,
            parentThreadId: options.parentThreadId,
        };
        return await service.createRequest(this.agentContext, createProofRequest);
    }
    async declineRequest(proofRecordId) {
        const proofRecord = await this.getById(proofRecordId);
        const service = this.getService(proofRecord.protocolVersion);
        proofRecord.assertState(ProofState_1.ProofState.RequestReceived);
        await service.updateState(this.agentContext, proofRecord, ProofState_1.ProofState.Declined);
        return proofRecord;
    }
    /**
     * Answer with a new presentation proposal in response to received presentation request message
     * to the connection associated with the proof record.
     *
     * @param options multiple properties like proof record id, proof format (indy/ presentation exchange)
     * to include in the message
     * @returns Proof record associated with the sent proposal message
     */
    async negotiateRequest(options) {
        const { proofRecordId } = options;
        const proofRecord = await this.getById(proofRecordId);
        const service = this.getService(proofRecord.protocolVersion);
        if (!proofRecord.connectionId) {
            throw new error_1.AriesFrameworkError(`No connectionId found for proof record '${proofRecord.id}'. Connection-less issuance does not support presentation proposal or negotiation.`);
        }
        const connection = await this.connectionService.getById(this.agentContext, proofRecord.connectionId);
        // Assert
        connection.assertReady();
        const proposalOptions = {
            proofRecord,
            proofFormats: options.proofFormats,
            autoAcceptProof: options.autoAcceptProof,
            goalCode: options.goalCode,
            comment: options.comment,
        };
        const { message } = await service.createProposalAsResponse(this.agentContext, proposalOptions);
        const outboundMessageContext = new models_1.OutboundMessageContext(message, {
            agentContext: this.agentContext,
            connection,
            associatedRecord: proofRecord,
        });
        await this.messageSender.sendMessage(outboundMessageContext);
        return proofRecord;
    }
    /**
     * Accept a presentation as prover (by sending a presentation acknowledgement message) to the connection
     * associated with the proof record.
     *
     * @param proofRecordId The id of the proof exchange record for which to accept the presentation
     * @returns Proof record associated with the sent presentation acknowledgement message
     *
     */
    async acceptPresentation(proofRecordId) {
        const record = await this.getById(proofRecordId);
        const service = this.getService(record.protocolVersion);
        const { message, proofRecord } = await service.createAck(this.agentContext, {
            proofRecord: record,
        });
        const requestMessage = await service.findRequestMessage(this.agentContext, record.id);
        const presentationMessage = await service.findPresentationMessage(this.agentContext, record.id);
        // Use connection if present
        if (proofRecord.connectionId) {
            const connection = await this.connectionService.getById(this.agentContext, proofRecord.connectionId);
            // Assert
            connection.assertReady();
            const outboundMessageContext = new models_1.OutboundMessageContext(message, {
                agentContext: this.agentContext,
                connection,
                associatedRecord: proofRecord,
            });
            await this.messageSender.sendMessage(outboundMessageContext);
        }
        // Use ~service decorator otherwise
        else if ((requestMessage === null || requestMessage === void 0 ? void 0 : requestMessage.service) && (presentationMessage === null || presentationMessage === void 0 ? void 0 : presentationMessage.service)) {
            const recipientService = presentationMessage === null || presentationMessage === void 0 ? void 0 : presentationMessage.service;
            const ourService = requestMessage.service;
            await this.messageSender.sendMessageToService(new models_1.OutboundMessageContext(message, {
                agentContext: this.agentContext,
                serviceParams: {
                    service: recipientService.resolvedDidCommService,
                    senderKey: ourService.resolvedDidCommService.recipientKeys[0],
                    returnRoute: true,
                },
            }));
        }
        // Cannot send message without credentialId or ~service decorator
        else {
            throw new error_1.AriesFrameworkError(`Cannot accept presentation without connectionId or ~service decorator on presentation message.`);
        }
        return record;
    }
    /**
     * Create a {@link RetrievedCredentials} object. Given input proof request and presentation proposal,
     * use credentials in the wallet to build indy requested credentials object for input to proof creation.
     * If restrictions allow, self attested attributes will be used.
     *
     * @param options multiple properties like proof record id and optional configuration
     * @returns RequestedCredentials
     */
    async autoSelectCredentialsForProofRequest(options) {
        const proofRecord = await this.getById(options.proofRecordId);
        const service = this.getService(proofRecord.protocolVersion);
        const retrievedCredentials = await service.getRequestedCredentialsForProofRequest(this.agentContext, {
            proofRecord: proofRecord,
            config: options.config,
        });
        return await service.autoSelectCredentialsForProofRequest(retrievedCredentials);
    }
    /**
     * Create a {@link RetrievedCredentials} object. Given input proof request and presentation proposal,
     * use credentials in the wallet to build indy requested credentials object for input to proof creation.
     *
     * If restrictions allow, self attested attributes will be used.
     *
     * @param options multiple properties like proof record id and optional configuration
     * @returns RetrievedCredentials object
     */
    async getRequestedCredentialsForProofRequest(options) {
        const record = await this.getById(options.proofRecordId);
        const service = this.getService(record.protocolVersion);
        return await service.getRequestedCredentialsForProofRequest(this.agentContext, {
            proofRecord: record,
            config: options.config,
        });
    }
    /**
     * Send problem report message for a proof record
     *
     * @param proofRecordId  The id of the proof record for which to send problem report
     * @param message message to send
     * @returns proof record associated with the proof problem report message
     */
    async sendProblemReport(proofRecordId, message) {
        const record = await this.getById(proofRecordId);
        const service = this.getService(record.protocolVersion);
        if (!record.connectionId) {
            throw new error_1.AriesFrameworkError(`No connectionId found for proof record '${record.id}'.`);
        }
        const connection = await this.connectionService.getById(this.agentContext, record.connectionId);
        // Assert
        connection.assertReady();
        const { message: problemReport } = await service.createProblemReport(this.agentContext, {
            proofRecord: record,
            description: message,
        });
        const outboundMessageContext = new models_1.OutboundMessageContext(problemReport, {
            agentContext: this.agentContext,
            connection,
            associatedRecord: record,
        });
        await this.messageSender.sendMessage(outboundMessageContext);
        return record;
    }
    async getFormatData(proofRecordId) {
        const proofRecord = await this.getById(proofRecordId);
        const service = this.getService(proofRecord.protocolVersion);
        return service.getFormatData(this.agentContext, proofRecordId);
    }
    /**
     * Retrieve all proof records
     *
     * @returns List containing all proof records
     */
    async getAll() {
        return this.proofRepository.getAll(this.agentContext);
    }
    /**
     * Retrieve all proof records by specified query params
     *
     * @returns List containing all proof records matching specified params
     */
    findAllByQuery(query) {
        return this.proofRepository.findByQuery(this.agentContext, query);
    }
    /**
     * Retrieve a proof record by id
     *
     * @param proofRecordId The proof record id
     * @throws {RecordNotFoundError} If no record is found
     * @return The proof record
     *
     */
    async getById(proofRecordId) {
        return await this.proofRepository.getById(this.agentContext, proofRecordId);
    }
    /**
     * Retrieve a proof record by id
     *
     * @param proofRecordId The proof record id
     * @return The proof record or null if not found
     *
     */
    async findById(proofRecordId) {
        return await this.proofRepository.findById(this.agentContext, proofRecordId);
    }
    /**
     * Delete a proof record by id
     *
     * @param proofId the proof record id
     */
    async deleteById(proofId, options) {
        const proofRecord = await this.getById(proofId);
        const service = this.getService(proofRecord.protocolVersion);
        return service.delete(this.agentContext, proofRecord, options);
    }
    /**
     * Retrieve a proof record by connection id and thread id
     *
     * @param connectionId The connection id
     * @param threadId The thread id
     * @throws {RecordNotFoundError} If no record is found
     * @throws {RecordDuplicateError} If multiple records are found
     * @returns The proof record
     */
    async getByThreadAndConnectionId(threadId, connectionId) {
        return this.proofRepository.getByThreadAndConnectionId(this.agentContext, threadId, connectionId);
    }
    /**
     * Retrieve proof records by connection id and parent thread id
     *
     * @param connectionId The connection id
     * @param parentThreadId The parent thread id
     * @returns List containing all proof records matching the given query
     */
    async getByParentThreadAndConnectionId(parentThreadId, connectionId) {
        return this.proofRepository.getByParentThreadAndConnectionId(this.agentContext, parentThreadId, connectionId);
    }
    /**
     * Update a proof record by
     *
     * @param proofRecord the proof record
     */
    async update(proofRecord) {
        await this.proofRepository.update(this.agentContext, proofRecord);
    }
    async findProposalMessage(proofRecordId) {
        const record = await this.getById(proofRecordId);
        const service = this.getService(record.protocolVersion);
        return service.findProposalMessage(this.agentContext, proofRecordId);
    }
    async findRequestMessage(proofRecordId) {
        const record = await this.getById(proofRecordId);
        const service = this.getService(record.protocolVersion);
        return service.findRequestMessage(this.agentContext, proofRecordId);
    }
    async findPresentationMessage(proofRecordId) {
        const record = await this.getById(proofRecordId);
        const service = this.getService(record.protocolVersion);
        return service.findPresentationMessage(this.agentContext, proofRecordId);
    }
    registerMessageHandlers(dispatcher, mediationRecipientService) {
        for (const service of Object.values(this.serviceMap)) {
            const proofService = service;
            proofService.registerMessageHandlers(dispatcher, this.agentConfig, new ProofResponseCoordinator_1.ProofResponseCoordinator(proofService), mediationRecipientService, this.routingService);
        }
    }
};
ProofsApi = __decorate([
    (0, tsyringe_1.injectable)(),
    __param(7, (0, tsyringe_1.inject)(constants_1.InjectionSymbols.Logger)),
    __metadata("design:paramtypes", [Dispatcher_1.Dispatcher,
        MediationRecipientService_1.MediationRecipientService,
        MessageSender_1.MessageSender,
        ConnectionService_1.ConnectionService,
        AgentContext_1.AgentContext,
        AgentConfig_1.AgentConfig,
        RoutingService_1.RoutingService, Object, ProofRepository_1.ProofRepository,
        V1ProofService_1.V1ProofService,
        V2ProofService_1.V2ProofService])
], ProofsApi);
exports.ProofsApi = ProofsApi;
