// @aries-framework/core/build/modules/dids/domain/keyDidDocument.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getDidDocumentForKey = void 0;
const crypto_1 = require("../../../crypto");
const constants_1 = require("../../vc/constants");
const constants_2 = require("../../vc/signature-suites/ed25519/constants");
const DidDocumentBuilder_1 = require("./DidDocumentBuilder");
const bls12381g1_1 = require("./key-type/bls12381g1");
const bls12381g1g2_1 = require("./key-type/bls12381g1g2");
const bls12381g2_1 = require("./key-type/bls12381g2");
const ed25519_1 = require("./key-type/ed25519");
const x25519_1 = require("./key-type/x25519");
const didDocumentKeyTypeMapping = {
    [crypto_1.KeyType.Ed25519]: getEd25519DidDoc,
    [crypto_1.KeyType.X25519]: getX25519DidDoc,
    [crypto_1.KeyType.Bls12381g1]: getBls12381g1DidDoc,
    [crypto_1.KeyType.Bls12381g2]: getBls12381g2DidDoc,
    [crypto_1.KeyType.Bls12381g1g2]: getBls12381g1g2DidDoc,
};
function getDidDocumentForKey(did, key) {
    const getDidDocument = didDocumentKeyTypeMapping[key.keyType];
    return getDidDocument(did, key);
}
exports.getDidDocumentForKey = getDidDocumentForKey;
function getBls12381g1DidDoc(did, key) {
    const verificationMethod = (0, bls12381g1_1.getBls12381g1VerificationMethod)(did, key);
    return getSignatureKeyBase({
        did,
        key,
        verificationMethod,
    })
        .addContext(constants_1.SECURITY_CONTEXT_BBS_URL)
        .build();
}
function getBls12381g1g2DidDoc(did, key) {
    const verificationMethods = (0, bls12381g1g2_1.getBls12381g1g2VerificationMethod)(did, key);
    const didDocumentBuilder = new DidDocumentBuilder_1.DidDocumentBuilder(did);
    for (const verificationMethod of verificationMethods) {
        didDocumentBuilder
            .addVerificationMethod(verificationMethod)
            .addAuthentication(verificationMethod.id)
            .addAssertionMethod(verificationMethod.id)
            .addCapabilityDelegation(verificationMethod.id)
            .addCapabilityInvocation(verificationMethod.id);
    }
    return didDocumentBuilder.addContext(constants_1.SECURITY_CONTEXT_BBS_URL).build();
}
function getEd25519DidDoc(did, key) {
    const verificationMethod = (0, ed25519_1.getEd25519VerificationMethod)({ id: `${did}#${key.fingerprint}`, key, controller: did });
    const publicKeyX25519 = (0, ed25519_1.convertPublicKeyToX25519)(key.publicKey);
    const didKeyX25519 = crypto_1.Key.fromPublicKey(publicKeyX25519, crypto_1.KeyType.X25519);
    const x25519VerificationMethod = (0, x25519_1.getX25519VerificationMethod)({
        id: `${did}#${didKeyX25519.fingerprint}`,
        key: didKeyX25519,
        controller: did,
    });
    const didDocBuilder = getSignatureKeyBase({ did, key, verificationMethod });
    didDocBuilder
        .addContext(constants_2.ED25519_SUITE_CONTEXT_URL_2018)
        .addContext(constants_1.SECURITY_X25519_CONTEXT_URL)
        .addKeyAgreement(x25519VerificationMethod);
    return didDocBuilder.build();
}
function getX25519DidDoc(did, key) {
    const verificationMethod = (0, x25519_1.getX25519VerificationMethod)({ id: `${did}#${key.fingerprint}`, key, controller: did });
    const document = new DidDocumentBuilder_1.DidDocumentBuilder(did)
        .addKeyAgreement(verificationMethod)
        .addContext(constants_1.SECURITY_X25519_CONTEXT_URL)
        .build();
    return document;
}
function getBls12381g2DidDoc(did, key) {
    const verificationMethod = (0, bls12381g2_1.getBls12381g2VerificationMethod)(did, key);
    return getSignatureKeyBase({
        did,
        key,
        verificationMethod,
    })
        .addContext(constants_1.SECURITY_CONTEXT_BBS_URL)
        .build();
}
function getSignatureKeyBase({ did, key, verificationMethod, }) {
    const keyId = `${did}#${key.fingerprint}`;
    return new DidDocumentBuilder_1.DidDocumentBuilder(did)
        .addVerificationMethod(verificationMethod)
        .addAuthentication(keyId)
        .addAssertionMethod(keyId)
        .addCapabilityDelegation(keyId)
        .addCapabilityInvocation(keyId);
}
