// indy-sdk-react-native/lib/index.js
/**
 * Copyright 2019 ABSA Group Limited
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */
import { NativeModules } from 'react-native';
import { Buffer } from 'buffer';
const nymRoleValues = {
  TRUSTEE: 0,
  STEWARD: 2,
  TRUST_ANCHOR: 101,
  ENDORSER: 101,
  NETWORK_MONITOR: 201
};
const {
  IndySdk
} = NativeModules;
const indy = {
  // wallet
  createWallet(config, credentials) {
    return IndySdk.createWallet(JSON.stringify(config), JSON.stringify(credentials));
  },

  openWallet(config, credentials) {
    return IndySdk.openWallet(JSON.stringify(config), JSON.stringify(credentials));
  },

  closeWallet(wh) {
    return IndySdk.closeWallet(wh);
  },

  deleteWallet(config, credentials) {
    return IndySdk.deleteWallet(JSON.stringify(config), JSON.stringify(credentials));
  },

  exportWallet(wh, exportConfig) {
    return IndySdk.exportWallet(wh, JSON.stringify(exportConfig));
  },

  importWallet(config, credentials, importConfig) {
    return IndySdk.importWallet(JSON.stringify(config), JSON.stringify(credentials), JSON.stringify(importConfig));
  },

  // did

  /**
   * Value of passed `wh` parameter will be ignored in Android version of Indy native bridge,
   * because it keeps wallet as a private attribute.
   */
  createAndStoreMyDid(wh, did) {
    if (Platform.OS === 'ios') {
      return IndySdk.createAndStoreMyDid(JSON.stringify(did), wh);
    }

    return IndySdk.createAndStoreMyDid(wh, JSON.stringify(did));
  },

  keyForDid(poolHandle, wh, did) {
    if (Platform.OS === 'ios') {
      return IndySdk.keyForDid(did, poolHandle, wh);
    }

    return IndySdk.keyForDid(poolHandle, wh, did);
  },

  keyForLocalDid(wh, did) {
    if (Platform.OS === 'ios') {
      return IndySdk.keyForLocalDid(did, wh);
    }

    return IndySdk.keyForLocalDid(wh, did);
  },

  storeTheirDid(wh, identity) {
    if (Platform.OS === 'ios') {
      return IndySdk.storeTheirDid(JSON.stringify(identity), wh);
    }

    throw new Error(`Unsupported operation! Platform: ${Platform.OS}`);
  },

  async listMyDidsWithMeta(wh) {
    if (Platform.OS === 'ios') {
      throw new Error(`Unsupported operation! Platform: ${Platform.OS}`);
    }

    return IndySdk.listMyDidsWithMeta(wh);
  },

  async setDidMetadata(wh, did, metadata) {
    if (Platform.OS === 'ios') {
      throw new Error(`Unsupported operation! Platform: ${Platform.OS}`);
    }

    return IndySdk.setDidMetadata(wh, did, metadata);
  },

  // pairwise
  createPairwise(wh, theirDid, myDid, metadata = '') {
    if (Platform.OS === 'ios') {
      return IndySdk.createPairwise(theirDid, myDid, metadata, wh);
    }

    return IndySdk.createPairwise(wh, theirDid, myDid, metadata);
  },

  async getPairwise(wh, theirDid) {
    if (Platform.OS === 'ios') {
      return JSON.parse((await IndySdk.getPairwise(theirDid, wh)));
    }

    return JSON.parse((await IndySdk.getPairwise(wh, theirDid)));
  },

  // crypto
  async createKey(wh, key) {
    if (Platform.OS === 'ios') {
      return IndySdk.createKey(JSON.stringify(key), wh);
    }

    return IndySdk.createKey(wh, JSON.stringify(key));
  },

  async cryptoAnonCrypt(recipientVk, messageRaw) {
    if (Platform.OS === 'ios') {
      return IndySdk.cryptoAnonCrypt(messageRaw, recipientVk);
    }

    return Buffer.from((await IndySdk.cryptoAnonCrypt(recipientVk, Array.from(messageRaw))));
  },

  async cryptoAnonDecrypt(wh, recipientVk, encryptedMsg) {
    if (Platform.OS === 'ios') {
      return IndySdk.cryptoAnonDecrypt(encryptedMsg, recipientVk, wh);
    }

    return Buffer.from((await IndySdk.cryptoAnonDecrypt(wh, recipientVk, Array.from(encryptedMsg))));
  },

  async cryptoAuthCrypt(wh, senderVk, recipientVk, messageRaw) {
    if (Platform.OS === 'ios') {
      return IndySdk.cryptoAuthCrypt(messageRaw, senderVk, recipientVk, wh);
    }

    return Buffer.from((await IndySdk.cryptoAuthCrypt(wh, senderVk, recipientVk, Array.from(messageRaw))));
  },

  async cryptoAuthDecrypt(wh, recipientVk, encryptedMsgRaw) {
    if (Platform.OS === 'ios') {
      return IndySdk.cryptoAuthDecrypt(encryptedMsgRaw, recipientVk, wh);
    }

    const [verkey, msg] = await IndySdk.cryptoAuthDecrypt(recipientVk, Array.from(encryptedMsgRaw));
    return [verkey, Buffer.from(msg)];
  },

  async cryptoSign(wh, signerVk, message) {
    return Buffer.from((await IndySdk.cryptoSign(wh, signerVk, Array.from(message))));
  },

  async cryptoVerify(signerVk, message, signature) {
    return IndySdk.cryptoVerify(signerVk, Array.from(message), Array.from(signature));
  },

  async packMessage(wh, message, receiverKeys, senderVk) {
    if (Platform.OS == 'ios') {
      return Buffer.from((await IndySdk.packMessage(wh, Array.from(message), JSON.stringify(receiverKeys), senderVk)));
    }

    return Buffer.from((await IndySdk.packMessage(wh, Array.from(message), receiverKeys, senderVk)));
  },

  async unpackMessage(wh, jwe) {
    return Buffer.from((await IndySdk.unpackMessage(wh, Array.from(jwe))));
  },

  // pool
  createPoolLedgerConfig(poolName, poolConfig) {
    return IndySdk.createPoolLedgerConfig(poolName, JSON.stringify(poolConfig));
  },

  openPoolLedger(poolName, poolConfig) {
    if (Platform.OS === 'ios') {
      if (poolConfig === undefined) {
        return IndySdk.openLedger(poolName, null);
      }

      return IndySdk.openLedger(poolName, JSON.stringify(poolConfig));
    }

    if (poolConfig === undefined) {
      return IndySdk.openPoolLedger(poolName, null);
    }

    return IndySdk.openPoolLedger(poolName, JSON.stringify(poolConfig));
  },

  setProtocolVersion(protocolVersion) {
    return IndySdk.setProtocolVersion(protocolVersion);
  },

  closePoolLedger(ph) {
    return IndySdk.closePoolLedger(ph);
  },

  // ledger
  async submitRequest(poolHandle, request) {
    if (Platform.OS === 'ios') {
      return JSON.parse((await IndySdk.submitRequest(JSON.stringify(request), poolHandle)));
    }

    return JSON.parse((await IndySdk.submitRequest(poolHandle, JSON.stringify(request))));
  },

  async signRequest(wh, submitterDid, request) {
    if (Platform.OS === 'ios') {
      throw new Error(`Unsupported operation! Platform: ${Platform.OS}`);
    }

    return JSON.parse((await IndySdk.signRequest(wh, submitterDid, JSON.stringify(request))));
  },

  async buildSchemaRequest(submitterDid, data) {
    if (Platform.OS === 'ios') {
      throw new Error(`Unsupported operation! Platform: ${Platform.OS}`);
    }

    return JSON.parse((await IndySdk.buildSchemaRequest(submitterDid, JSON.stringify(data))));
  },

  async buildGetSchemaRequest(submitterDid, id) {
    return JSON.parse((await IndySdk.buildGetSchemaRequest(submitterDid, id)));
  },

  async buildGetTxnRequest(submitterDid, ledgerType, seqNo) {
    return JSON.parse((await IndySdk.buildGetTxnRequest(submitterDid, ledgerType, seqNo)));
  },

  async buildGetAttribRequest(submitterDid, targetDid, raw, hash, enc) {
    return JSON.parse((await IndySdk.buildGetAttribRequest(submitterDid, targetDid, raw, hash, enc)));
  },

  async buildGetNymRequest(submitterDid, targetDid) {
    return JSON.parse((await IndySdk.buildGetNymRequest(submitterDid, targetDid)));
  },

  async parseGetNymResponse(response) {
    return JSON.parse((await IndySdk.parseGetNymResponse(JSON.stringify(response))));
  },

  async parseGetSchemaResponse(getSchemaResponse) {
    const [id, schema] = await IndySdk.parseGetSchemaResponse(JSON.stringify(getSchemaResponse));
    return [id, JSON.parse(schema)];
  },

  async buildCredDefRequest(submitterDid, credDef) {
    if (Platform.OS === 'ios') {
      throw new Error(`Unsupported operation! Platform: ${Platform.OS}`);
    }

    return JSON.parse((await IndySdk.buildCredDefRequest(submitterDid, JSON.stringify(credDef))));
  },

  async buildGetCredDefRequest(submitterDid, id) {
    return JSON.parse((await IndySdk.buildGetCredDefRequest(submitterDid, id)));
  },

  async parseGetCredDefResponse(getCredDefResponse) {
    const [credDefId, credDef] = await IndySdk.parseGetCredDefResponse(JSON.stringify(getCredDefResponse));
    return [credDefId, JSON.parse(credDef)];
  },

  async buildGetRevocRegDefRequest(submitterDid, revocRegDefId) {
    return JSON.parse((await IndySdk.buildGetRevocRegDefRequest(submitterDid, revocRegDefId)));
  },

  async parseGetRevocRegDefResponse(getRevocRegResponse) {
    const [revocRegDefId, revocRegDef] = await IndySdk.parseGetRevocRegDefResponse(JSON.stringify(getRevocRegResponse));
    return [revocRegDefId, JSON.parse(revocRegDef)];
  },

  async buildGetRevocRegDeltaRequest(submitterDid, revocRegDefId, from = 0, to = new Date().getTime()) {
    return JSON.parse((await IndySdk.buildGetRevocRegDeltaRequest(submitterDid, revocRegDefId, from, to)));
  },

  async parseGetRevocRegDeltaResponse(getRevocRegDeltaResponse) {
    const [revocRegId, revocRegDelta, timestamp] = await IndySdk.parseGetRevocRegDeltaResponse(JSON.stringify(getRevocRegDeltaResponse));
    return [revocRegId, JSON.parse(revocRegDelta), timestamp];
  },

  async buildGetRevocRegRequest(submitterDid, revocRegDefId, timestamp) {
    return JSON.parse((await IndySdk.buildGetRevocRegRequest(submitterDid, revocRegDefId, timestamp)));
  },

  async parseGetRevocRegResponse(getRevocRegResponse) {
    const [revocRegId, revocReg, ledgerTimestamp] = await IndySdk.parseGetRevocRegResponse(JSON.stringify(getRevocRegResponse));
    return [revocRegId, JSON.parse(revocReg), ledgerTimestamp];
  },

  async proverCreateMasterSecret(wh, masterSecretId) {
    if (Platform.OS === 'ios') {
      return IndySdk.proverCreateMasterSecret(masterSecretId, wh);
    }

    return IndySdk.proverCreateMasterSecret(wh, masterSecretId);
  },

  async proverCreateCredentialReq(wh, proverDid, credOffer, credDef, masterSecretId) {
    if (Platform.OS === 'ios') {
      const [credReq, credReqMetadata] = await IndySdk.proverCreateCredentialReq(JSON.stringify(credOffer), JSON.stringify(credDef), proverDid, masterSecretId, wh);
      return [JSON.parse(credReq), JSON.parse(credReqMetadata)];
    }

    const [credReq, credReqMetadata] = await IndySdk.proverCreateCredentialReq(wh, proverDid, JSON.stringify(credOffer), JSON.stringify(credDef), masterSecretId);
    return [JSON.parse(credReq), JSON.parse(credReqMetadata)];
  },

  proverStoreCredential(wh, credId, credReqMetadata, cred, credDef, revRegDef) {
    if (Platform.OS === 'ios') {
      return IndySdk.proverStoreCredential(JSON.stringify(cred), credId, JSON.stringify(credReqMetadata), JSON.stringify(credDef), revRegDef && JSON.stringify(revRegDef), wh);
    }

    return IndySdk.proverStoreCredential(wh, credId, JSON.stringify(credReqMetadata), JSON.stringify(cred), JSON.stringify(credDef), revRegDef && JSON.stringify(revRegDef));
  },

  async proverGetCredential(wh, credId) {
    if (Platform.OS === 'ios') {
      return JSON.parse((await IndySdk.proverGetCredential(credId, wh)));
    }

    return JSON.parse((await IndySdk.proverGetCredential(wh, credId)));
  },

  async proverDeleteCredential(wh, credId) {
    if (Platform.OS === 'ios') {
      return await IndySdk.proverDeleteCredential(credId, wh);
    }

    return await IndySdk.proverDeleteCredential(wh, credId);
  },

  // NOTE: This method is deprecated because immediately returns all fetched credentials. Use proverSearchCredentials() to fetch records by small batches.
  async proverGetCredentials(wh, filter = {}) {
    if (Platform.OS === 'ios') {
      return JSON.parse((await IndySdk.proverGetCredentials(JSON.stringify(filter), wh)));
    }

    return JSON.parse((await IndySdk.proverGetCredentials(wh, JSON.stringify(filter))));
  },

  // TODO: add proverSearchCredentials() method
  // TODO: add proverFetchCredentials() method
  // TODO: add proverCloseCredentialsSearch() method
  // TODO Add return flow type.
  // It needs little investigation, because is doesn't seem to be same format as Credential stored in wallet.
  // NOTE: This method is deprecated because immediately returns all fetched credentials. proverSearchCredentialsForProofReq to fetch records by small batches.
  async proverGetCredentialsForProofReq(wh, proofRequest) {
    if (Platform.OS == 'ios') {
      return JSON.parse((await IndySdk.proverGetCredentialsForProofReq(JSON.stringify(proofRequest), wh)));
    }

    return JSON.parse((await IndySdk.proverGetCredentialsForProofReq(wh, JSON.stringify(proofRequest))));
  },

  async proverSearchCredentialsForProofReq(wh, proofRequest, extraQuery) {
    //The NodeJS IndySDK wrapper differs from the Java and iOS wrappers in this call--it allows extraQuery to be a wql query object or null.
    return await IndySdk.proverSearchCredentialsForProofReq(wh, JSON.stringify(proofRequest), JSON.stringify(extraQuery ?? {}));
  },

  async proverFetchCredentialsForProofReq(sh, itemReferent, count) {
    return JSON.parse((await IndySdk.proverFetchCredentialsForProofReq(sh, itemReferent, count)));
  },

  async proverCloseCredentialsSearchForProofReq(sh) {
    return await IndySdk.proverCloseCredentialsSearchForProofReq(sh);
  },

  async proverCreateProof(wh, proofReq, requestedCredentials, masterSecretName, schemas, credentialDefs, revStates = {}) {
    if (Platform.OS == 'ios') {
      return JSON.parse((await IndySdk.proverCreateProofForRequest(JSON.stringify(proofReq), JSON.stringify(requestedCredentials), masterSecretName, JSON.stringify(schemas), JSON.stringify(credentialDefs), JSON.stringify(revStates), wh)));
    }

    return JSON.parse((await IndySdk.proverCreateProof(wh, JSON.stringify(proofReq), JSON.stringify(requestedCredentials), masterSecretName, JSON.stringify(schemas), JSON.stringify(credentialDefs), JSON.stringify(revStates))));
  },

  async verifierVerifyProof(proofRequest, proof, schemas, credentialDefs, revRegDefs, revStates) {
    return IndySdk.verifierVerifyProof(JSON.stringify(proofRequest), JSON.stringify(proof), JSON.stringify(schemas), JSON.stringify(credentialDefs), JSON.stringify(revRegDefs), JSON.stringify(revStates));
  },

  async generateNonce() {
    return IndySdk.generateNonce();
  },

  async generateWalletKey(config = {}) {
    return IndySdk.generateWalletKey(JSON.stringify(config));
  },

  async appendTxnAuthorAgreementAcceptanceToRequest(request, text, version, taaDigest, mechanism, time) {
    if (Platform.OS === 'ios') {
      throw new Error(`Unsupported operation! Platform: ${Platform.OS}`);
    }

    return JSON.parse((await IndySdk.appendTxnAuthorAgreementAcceptanceToRequest(JSON.stringify(request), text, version, taaDigest, mechanism, time)));
  },

  async buildGetTxnAuthorAgreementRequest(submitterDid, data) {
    if (Platform.OS === 'ios') {
      throw new Error(`Unsupported operation! Platform: ${Platform.OS}`);
    }

    return JSON.parse((await IndySdk.buildGetTxnAuthorAgreementRequest(submitterDid, data)));
  },

  // non_secrets
  async addWalletRecord(wh, type, id, value, tags) {
    return IndySdk.addWalletRecord(wh, type, id, value, JSON.stringify(tags));
  },

  async updateWalletRecordValue(wh, type, id, value) {
    return IndySdk.updateWalletRecordValue(wh, type, id, value);
  },

  async updateWalletRecordTags(wh, type, id, tags) {
    return IndySdk.updateWalletRecordTags(wh, type, id, JSON.stringify(tags));
  },

  async addWalletRecordTags(wh, type, id, tags) {
    return IndySdk.addWalletRecordTags(wh, type, id, JSON.stringify(tags));
  },

  async deleteWalletRecordTags(wh, type, id, tagNames) {
    return IndySdk.deleteWalletRecordTags(wh, type, id, JSON.stringify(tagNames));
  },

  async deleteWalletRecord(wh, type, id) {
    return IndySdk.deleteWalletRecord(wh, type, id);
  },

  async getWalletRecord(wh, type, id, options) {
    return JSON.parse((await IndySdk.getWalletRecord(wh, type, id, JSON.stringify(options))));
  },

  async openWalletSearch(wh, type, query, options) {
    return IndySdk.openWalletSearch(wh, type, JSON.stringify(query), JSON.stringify(options));
  },

  async fetchWalletSearchNextRecords(wh, sh, count) {
    return JSON.parse((await IndySdk.fetchWalletSearchNextRecords(wh, sh, count)));
  },

  async closeWalletSearch(sh) {
    return IndySdk.closeWalletSearch(sh);
  },

  // Anoncreds
  async issuerCreateSchema(did, name, version, attributes) {
    if (Platform.OS === 'ios') {
      throw new Error(`Unsupported operation! Platform: ${Platform.OS}`);
    }

    const [schemaId, schema] = await IndySdk.issuerCreateSchema(did, name, version, JSON.stringify(attributes));
    return [schemaId, JSON.parse(schema)];
  },

  async issuerCreateAndStoreCredentialDef(wh, issuerDid, schema, tag, signatureType, config) {
    if (Platform.OS === 'ios') {
      throw new Error(`Unsupported operation! Platform: ${Platform.OS}`);
    }

    const [credDefId, credDef] = await IndySdk.issuerCreateAndStoreCredentialDef(wh, issuerDid, JSON.stringify(schema), tag, signatureType, JSON.stringify(config));
    return [credDefId, JSON.parse(credDef)];
  },

  async issuerCreateCredentialOffer(wh, credDefId) {
    if (Platform.OS === 'ios') {
      throw new Error(`Unsupported operation! Platform: ${Platform.OS}`);
    }

    return JSON.parse((await IndySdk.issuerCreateCredentialOffer(wh, credDefId)));
  },

  async issuerCreateCredential(wh, credOffer, credReq, credvalues, revRegId, blobStorageReaderHandle) {
    if (Platform.OS === 'ios') {
      throw new Error(`Unsupported operation! Platform: ${Platform.OS}`);
    }

    const [credJson, revocId, revocRegDelta] = await IndySdk.issuerCreateCredential(wh, JSON.stringify(credOffer), JSON.stringify(credReq), JSON.stringify(credvalues), revRegId, blobStorageReaderHandle);
    return [JSON.parse(credJson), revocId, JSON.parse(revocRegDelta)];
  },

  async createRevocationState(blobStorageReaderHandle, revRegDef, revRegDelta, timestamp, credRevId) {
    return JSON.parse((await IndySdk.createRevocationState(blobStorageReaderHandle, JSON.stringify(revRegDef), JSON.stringify(revRegDelta), timestamp, credRevId)));
  },

  // blob_storage
  async openBlobStorageReader(type, tailsWriterConfig) {
    return JSON.parse((await IndySdk.openBlobStorageReader(type, JSON.stringify(tailsWriterConfig))));
  }

};
const indyErrors = {
  100: 'CommonInvalidParam1',
  101: 'CommonInvalidParam2',
  102: 'CommonInvalidParam3',
  103: 'CommonInvalidParam4',
  104: 'CommonInvalidParam5',
  105: 'CommonInvalidParam6',
  106: 'CommonInvalidParam7',
  107: 'CommonInvalidParam8',
  108: 'CommonInvalidParam9',
  109: 'CommonInvalidParam10',
  110: 'CommonInvalidParam11',
  111: 'CommonInvalidParam12',
  112: 'CommonInvalidState',
  113: 'CommonInvalidStructure',
  114: 'CommonIOError',
  115: 'CommonInvalidParam13',
  116: 'CommonInvalidParam14',
  200: 'WalletInvalidHandle',
  201: 'WalletUnknownTypeError',
  202: 'WalletTypeAlreadyRegisteredError',
  203: 'WalletAlreadyExistsError',
  204: 'WalletNotFoundError',
  205: 'WalletIncompatiblePoolError',
  206: 'WalletAlreadyOpenedError',
  207: 'WalletAccessFailed',
  208: 'WalletInputError',
  209: 'WalletDecodingError',
  210: 'WalletStorageError',
  211: 'WalletEncryptionError',
  212: 'WalletItemNotFound',
  213: 'WalletItemAlreadyExists',
  214: 'WalletQueryError',
  300: 'PoolLedgerNotCreatedError',
  301: 'PoolLedgerInvalidPoolHandle',
  302: 'PoolLedgerTerminated',
  303: 'LedgerNoConsensusError',
  304: 'LedgerInvalidTransaction',
  305: 'LedgerSecurityError',
  306: 'PoolLedgerConfigAlreadyExistsError',
  307: 'PoolLedgerTimeout',
  308: 'PoolIncompatibleProtocolVersion',
  309: 'LedgerNotFound',
  400: 'AnoncredsRevocationRegistryFullError',
  401: 'AnoncredsInvalidUserRevocId',
  404: 'AnoncredsMasterSecretDuplicateNameError',
  405: 'AnoncredsProofRejected',
  406: 'AnoncredsCredentialRevoked',
  407: 'AnoncredsCredDefAlreadyExistsError',
  500: 'UnknownCryptoTypeError',
  600: 'DidAlreadyExistsError',
  700: 'PaymentUnknownMethodError',
  701: 'PaymentIncompatibleMethodsError',
  702: 'PaymentInsufficientFundsError',
  703: 'PaymentSourceDoesNotExistError',
  704: 'PaymentOperationNotSupportedError',
  705: 'PaymentExtraFundsError',
  706: 'TransactionNotAllowedError'
};

function wrapIndyCallWithErrorHandling(func) {
  return async (...args) => {
    // Wrap try/catch around indy func
    try {
      return await func(...args);
    } catch (e) {
      // Try to parse e.message as it should be a
      // JSON string for Indy errors (at least on Android)
      // Parsing could also go wrong. In this case we just
      // want to throw the native side error
      let parse;

      try {
        parse = JSON.parse(e.message);
      } catch {
        throw e;
      }

      if (Platform.OS === 'ios') {
        const {
          indyCode,
          message
        } = parse;

        if (!isNaN(indyCode) && indyErrors.hasOwnProperty(indyCode)) {
          const indyName = indyErrors[indyCode];
          const indyErrorFromIOS = {
            name: 'IndyError',
            message,
            indyCode: indyCode,
            indyName: indyName,
            indyCurrentErrorJson: null
          };
          throw indyErrorFromIOS;
        }
      }

      throw parse;
    }
  };
} // This adds indy error handling to all methods to
// transform the string messages into JSON error objects


const indyWithErrorHandling = Object.fromEntries(Object.entries(indy).map(([funcName, funcImpl]) => [funcName, wrapIndyCallWithErrorHandling(funcImpl)]));
export default indyWithErrorHandling;
