// @aries-framework/core/build/modules/vc/models/credential/Issuer.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.IsIssuer = exports.IssuerTransformer = exports.Issuer = void 0;
const class_transformer_1 = require("class-transformer");
const class_validator_1 = require("class-validator");
const validators_1 = require("../../../../utils/validators");
class Issuer {
    constructor(options) {
        if (options) {
            this.id = options.id;
        }
    }
}
__decorate([
    (0, validators_1.IsUri)(),
    __metadata("design:type", String)
], Issuer.prototype, "id", void 0);
exports.Issuer = Issuer;
// Custom transformers
function IssuerTransformer() {
    return (0, class_transformer_1.Transform)(({ value, type }) => {
        if (type === class_transformer_1.TransformationType.PLAIN_TO_CLASS) {
            if ((0, class_validator_1.isString)(value))
                return value;
            return (0, class_transformer_1.plainToInstance)(Issuer, value);
        }
        else if (type === class_transformer_1.TransformationType.CLASS_TO_PLAIN) {
            if ((0, class_validator_1.isString)(value))
                return value;
            return (0, class_transformer_1.instanceToPlain)(value);
        }
        // PLAIN_TO_PLAIN
        return value;
    });
}
exports.IssuerTransformer = IssuerTransformer;
// Custom validators
function IsIssuer(validationOptions) {
    return (0, class_validator_1.ValidateBy)({
        name: 'IsIssuer',
        validator: {
            validate: (value) => {
                if (typeof value === 'string') {
                    return validators_1.UriValidator.test(value);
                }
                if ((0, class_validator_1.isInstance)(value, Issuer)) {
                    return validators_1.UriValidator.test(value.id);
                }
                return false;
            },
            defaultMessage: (0, class_validator_1.buildMessage)((eachPrefix) => eachPrefix + '$property must be a string or an object with an id property', validationOptions),
        },
    }, validationOptions);
}
exports.IsIssuer = IsIssuer;
