// @aries-framework/core/build/modules/vc/models/credential/W3cCredential.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.IsCredentialType = exports.W3cCredential = void 0;
const class_transformer_1 = require("class-transformer");
const class_validator_1 = require("class-validator");
const validators_1 = require("../../../../utils/validators");
const constants_1 = require("../../constants");
const validators_2 = require("../../validators");
const CredentialSchema_1 = require("./CredentialSchema");
const CredentialSubject_1 = require("./CredentialSubject");
const Issuer_1 = require("./Issuer");
class W3cCredential {
    constructor(options) {
        var _a;
        if (options) {
            this.context = (_a = options.context) !== null && _a !== void 0 ? _a : [constants_1.CREDENTIALS_CONTEXT_V1_URL];
            this.id = options.id;
            this.type = options.type || [];
            this.issuer = options.issuer;
            this.issuanceDate = options.issuanceDate;
            this.expirationDate = options.expirationDate;
            this.credentialSubject = options.credentialSubject;
        }
    }
    get issuerId() {
        return this.issuer instanceof Issuer_1.Issuer ? this.issuer.id : this.issuer;
    }
    get credentialSchemaIds() {
        if (!this.credentialSchema)
            return [];
        if (Array.isArray(this.credentialSchema)) {
            return this.credentialSchema.map((credentialSchema) => credentialSchema.id);
        }
        return [this.credentialSchema.id];
    }
    get credentialSubjectIds() {
        if (Array.isArray(this.credentialSubject)) {
            return this.credentialSubject.map((credentialSubject) => credentialSubject.id);
        }
        return [this.credentialSubject.id];
    }
    get contexts() {
        if (Array.isArray(this.context)) {
            return this.context.filter((x) => typeof x === 'string');
        }
        if (typeof this.context === 'string') {
            return [this.context];
        }
        return [this.context.id];
    }
}
__decorate([
    (0, class_transformer_1.Expose)({ name: '@context' }),
    (0, validators_2.IsJsonLdContext)(),
    __metadata("design:type", Object)
], W3cCredential.prototype, "context", void 0);
__decorate([
    (0, class_validator_1.IsOptional)(),
    (0, validators_1.IsUri)(),
    __metadata("design:type", String)
], W3cCredential.prototype, "id", void 0);
__decorate([
    IsCredentialType(),
    __metadata("design:type", Array)
], W3cCredential.prototype, "type", void 0);
__decorate([
    (0, Issuer_1.IssuerTransformer)(),
    (0, Issuer_1.IsIssuer)(),
    __metadata("design:type", Object)
], W3cCredential.prototype, "issuer", void 0);
__decorate([
    (0, class_validator_1.IsString)(),
    __metadata("design:type", String)
], W3cCredential.prototype, "issuanceDate", void 0);
__decorate([
    (0, class_validator_1.IsString)(),
    (0, class_validator_1.IsOptional)(),
    __metadata("design:type", String)
], W3cCredential.prototype, "expirationDate", void 0);
__decorate([
    (0, class_transformer_1.Type)(() => CredentialSubject_1.CredentialSubject),
    (0, validators_1.IsInstanceOrArrayOfInstances)({ classType: CredentialSubject_1.CredentialSubject }),
    __metadata("design:type", Object)
], W3cCredential.prototype, "credentialSubject", void 0);
__decorate([
    (0, class_validator_1.IsOptional)(),
    (0, class_transformer_1.Type)(() => CredentialSchema_1.CredentialSchema),
    (0, validators_1.IsInstanceOrArrayOfInstances)({ classType: CredentialSchema_1.CredentialSchema }),
    __metadata("design:type", Object)
], W3cCredential.prototype, "credentialSchema", void 0);
exports.W3cCredential = W3cCredential;
// Custom validator
function IsCredentialType(validationOptions) {
    return (0, class_validator_1.ValidateBy)({
        name: 'IsVerifiableCredentialType',
        validator: {
            validate: (value) => {
                if (Array.isArray(value)) {
                    return value.includes(constants_1.VERIFIABLE_CREDENTIAL_TYPE);
                }
                return false;
            },
            defaultMessage: (0, class_validator_1.buildMessage)((eachPrefix) => eachPrefix + '$property must be an array of strings which includes "VerifiableCredential"', validationOptions),
        },
    }, validationOptions);
}
exports.IsCredentialType = IsCredentialType;
