// @aries-framework/core/build/modules/proofs/protocol/v1/handlers/V1RequestPresentationHandler.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.V1RequestPresentationHandler = void 0;
const models_1 = require("../../../../../agent/models");
const ServiceDecorator_1 = require("../../../../../decorators/service/ServiceDecorator");
const error_1 = require("../../../../../error");
const storage_1 = require("../../../../../storage");
const messages_1 = require("../messages");
class V1RequestPresentationHandler {
    constructor(proofService, agentConfig, proofResponseCoordinator, mediationRecipientService, didCommMessageRepository, routingService) {
        this.supportedMessages = [messages_1.V1RequestPresentationMessage];
        this.proofService = proofService;
        this.agentConfig = agentConfig;
        this.proofResponseCoordinator = proofResponseCoordinator;
        this.mediationRecipientService = mediationRecipientService;
        this.didCommMessageRepository = didCommMessageRepository;
        this.routingService = routingService;
    }
    async handle(messageContext) {
        const proofRecord = await this.proofService.processRequest(messageContext);
        const shouldAutoRespond = await this.proofResponseCoordinator.shouldAutoRespondToRequest(messageContext.agentContext, proofRecord);
        if (shouldAutoRespond) {
            return await this.createPresentation(proofRecord, messageContext);
        }
    }
    async createPresentation(record, messageContext) {
        const requestMessage = await this.didCommMessageRepository.getAgentMessage(messageContext.agentContext, {
            associatedRecordId: record.id,
            messageClass: messages_1.V1RequestPresentationMessage,
        });
        const indyProofRequest = requestMessage.indyProofRequest;
        this.agentConfig.logger.info(`Automatically sending presentation with autoAccept on ${this.agentConfig.autoAcceptProofs}`);
        if (!indyProofRequest) {
            this.agentConfig.logger.error('Proof request is undefined.');
            throw new error_1.AriesFrameworkError('No proof request found.');
        }
        const retrievedCredentials = await this.proofService.getRequestedCredentialsForProofRequest(messageContext.agentContext, {
            proofRecord: record,
            config: {
                filterByPresentationPreview: true,
            },
        });
        if (!retrievedCredentials.proofFormats.indy) {
            this.agentConfig.logger.error('No matching Indy credentials could be retrieved.');
            throw new error_1.AriesFrameworkError('No matching Indy credentials could be retrieved.');
        }
        const options = {
            proofFormats: retrievedCredentials.proofFormats,
        };
        const requestedCredentials = await this.proofService.autoSelectCredentialsForProofRequest(options);
        const { message, proofRecord } = await this.proofService.createPresentation(messageContext.agentContext, {
            proofRecord: record,
            proofFormats: {
                indy: requestedCredentials.proofFormats.indy,
            },
            willConfirm: true,
        });
        if (messageContext.connection) {
            return new models_1.OutboundMessageContext(message, {
                agentContext: messageContext.agentContext,
                connection: messageContext.connection,
                associatedRecord: proofRecord,
            });
        }
        else if (requestMessage.service) {
            const routing = await this.routingService.getRouting(messageContext.agentContext);
            message.service = new ServiceDecorator_1.ServiceDecorator({
                serviceEndpoint: routing.endpoints[0],
                recipientKeys: [routing.recipientKey.publicKeyBase58],
                routingKeys: routing.routingKeys.map((key) => key.publicKeyBase58),
            });
            const recipientService = requestMessage.service;
            await this.didCommMessageRepository.saveOrUpdateAgentMessage(messageContext.agentContext, {
                agentMessage: message,
                associatedRecordId: proofRecord.id,
                role: storage_1.DidCommMessageRole.Sender,
            });
            return new models_1.OutboundMessageContext(message, {
                agentContext: messageContext.agentContext,
                serviceParams: {
                    service: recipientService.resolvedDidCommService,
                    senderKey: message.service.resolvedDidCommService.recipientKeys[0],
                },
            });
        }
        this.agentConfig.logger.error(`Could not automatically create presentation`);
    }
}
exports.V1RequestPresentationHandler = V1RequestPresentationHandler;
